{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Скачивание файлов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "base = \"http://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/\"\n",
    "paths = {\n",
    "    \"E_billingiae\": \"000/196/615/GCF_000196615.1_ASM19661v1\",\n",
    "    \"E_gerundensis\": \"020/342/335/GCF_020342335.1_ASM2034233v1\",\n",
    "    \"E_persicina\": \"019/844/095/GCF_019844095.1_ASM1984409v1\",\n",
    "    \"E_pyrifoliae\": \"002/952/315/GCF_002952315.1_ASM295231v1\",\n",
    "    \"E_tasmaniensis\": \"000/026/185/GCF_000026185.1_ASM2618v1\"\n",
    "}\n",
    "genomic_end = \"_genomic.fna.gz\"\n",
    "feature_end = \"_feature_table.txt.gz\"\n",
    "protein_end = \"_protein.faa.gz\"\n",
    "\n",
    "! mkdir -p genomes histograms zhunt_result TSS intersect ortho"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in paths.keys():\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{genomic_end} -O 'genomes/{key}{genomic_end}'\")\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{feature_end} -O 'genomes/{key}{feature_end}'\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{genomic_end}\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{feature_end}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in paths.keys():\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{protein_end} -O 'genomes/{key}{protein_end}'\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{protein_end}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Анализ аннотированных геномов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_list = list(paths.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annot_analyse(genome):\n",
    "    features = pd.read_csv(f\"genomes/{genome}_feature_table.txt\", sep='\\t')\n",
    "    features = features.loc[\n",
    "        (features['# feature']=='gene') & \n",
    "        (features['seq_type']=='chromosome')\n",
    "    ]\n",
    "    seqs = [len(seq_record.seq) for seq_record in SeqIO.parse(f\"genomes/{genome}_genomic.fna\", \"fasta\")]\n",
    "    # Choose complete genome, remove plasmids\n",
    "    genome_len = max(seqs)\n",
    "\n",
    "    total_genes_len = features[\"feature_interval_length\"].sum()\n",
    "\n",
    "    return [len(features), round(total_genes_len/genome_len*100, 2), genome_len]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_billingiae</th>\n",
       "      <th>E_gerundensis</th>\n",
       "      <th>E_persicina</th>\n",
       "      <th>E_pyrifoliae</th>\n",
       "      <th>E_tasmaniensis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Genes number</th>\n",
       "      <td>4703.00</td>\n",
       "      <td>3458.00</td>\n",
       "      <td>4311.00</td>\n",
       "      <td>3746.00</td>\n",
       "      <td>3569.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome coverage</th>\n",
       "      <td>88.99</td>\n",
       "      <td>88.19</td>\n",
       "      <td>88.61</td>\n",
       "      <td>86.27</td>\n",
       "      <td>86.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome length</th>\n",
       "      <td>5100167.00</td>\n",
       "      <td>3748909.00</td>\n",
       "      <td>4664605.00</td>\n",
       "      <td>4027225.00</td>\n",
       "      <td>3883467.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 E_billingiae  E_gerundensis  E_persicina  E_pyrifoliae  \\\n",
       "Genes number          4703.00        3458.00      4311.00       3746.00   \n",
       "Genome coverage         88.99          88.19        88.61         86.27   \n",
       "Genome length      5100167.00     3748909.00   4664605.00    4027225.00   \n",
       "\n",
       "                 E_tasmaniensis  \n",
       "Genes number            3569.00  \n",
       "Genome coverage           86.71  \n",
       "Genome length        3883467.00  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    annot_dict[genome] = annot_analyse(genome)\n",
    "\n",
    "pd.DataFrame(annot_dict, index=[\"Genes number\", \"Genome coverage\", \"Genome length\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Предсказание участков Z-DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system(\"wget -c -nv https://raw.githubusercontent.com/vanya-antonov/hse22-project/main/zhunt3-alan.c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system(\"gcc zhunt3-alan.c -lm -o zhunt3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from subprocess import DEVNULL, PIPE\n",
    "\n",
    "ZH_EXECUTABLE = Path(\"../hse22_project/zhunt3\")\n",
    "assert ZH_EXECUTABLE.is_file()\n",
    "\n",
    "\n",
    "def zhunt(query: str, windowsize: int = 6, minsize: int = 3, maxsize: int = 6):\n",
    "    assert set(query).issubset({\"A\", \"C\", \"G\", \"T\", \"N\"})\n",
    "    fd, temp = tempfile.mkstemp()\n",
    "    os.close(fd)\n",
    "    with open(temp, 'w') as stream:\n",
    "        stream.write(query)\n",
    "\n",
    "    subprocess.run(\n",
    "        [ZH_EXECUTABLE, \n",
    "         str(windowsize), str(minsize), str(maxsize), temp],\n",
    "        check=True, stdout=PIPE, stderr=DEVNULL,\n",
    "        input=query, encoding='ascii'\n",
    "    )\n",
    "    with open(temp + \".Z-SCORE\", 'r') as stream:\n",
    "        df = pd.read_csv(stream,\n",
    "                         names=['Start', 'End', 'nu-1', 'nu-2', 'nu-3', \n",
    "                                'ZH-Score', 'Sequence', 'Conformation'],\n",
    "                         skiprows=1, sep='\\s+')\n",
    "    os.remove(temp)\n",
    "    os.remove(temp + \".Z-SCORE\")\n",
    "    df = df.loc[df[\"ZH-Score\"] > 500]\n",
    "    return df[['Start', 'End', 'ZH-Score', 'Sequence', 'Conformation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def choose_max_seq(genome):\n",
    "    seqs = [str(seq_record.seq) for seq_record in SeqIO.parse(f\"genomes/{genome}_genomic.fna\", \"fasta\")]\n",
    "    # Choose complete genome, remove plasmids\n",
    "    max_scaffold = max(seqs, key=len)\n",
    "    return max_scaffold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "zhunt_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    zhunt_dict[genome] = zhunt(choose_max_seq(genome))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ZH-Score</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Conformation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>734</td>\n",
       "      <td>742</td>\n",
       "      <td>766.6232</td>\n",
       "      <td>cgtgcgtg</td>\n",
       "      <td>ASASASAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1677</th>\n",
       "      <td>1678</td>\n",
       "      <td>1690</td>\n",
       "      <td>1356.2120</td>\n",
       "      <td>gggcgtgcccgc</td>\n",
       "      <td>SASASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1679</th>\n",
       "      <td>1680</td>\n",
       "      <td>1690</td>\n",
       "      <td>1201.6710</td>\n",
       "      <td>gcgtgcccgc</td>\n",
       "      <td>SASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>1745</td>\n",
       "      <td>1753</td>\n",
       "      <td>516.0707</td>\n",
       "      <td>ttgcgcgc</td>\n",
       "      <td>SASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1746</th>\n",
       "      <td>1747</td>\n",
       "      <td>1753</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>gcgcgc</td>\n",
       "      <td>SASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5096951</th>\n",
       "      <td>5096952</td>\n",
       "      <td>5096962</td>\n",
       "      <td>2183.5740</td>\n",
       "      <td>cgtgcgtgtg</td>\n",
       "      <td>ASASASASAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5097858</th>\n",
       "      <td>5097859</td>\n",
       "      <td>5097871</td>\n",
       "      <td>717.9641</td>\n",
       "      <td>aagcgtatgcgc</td>\n",
       "      <td>SASASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5097860</th>\n",
       "      <td>5097861</td>\n",
       "      <td>5097871</td>\n",
       "      <td>948.8341</td>\n",
       "      <td>gcgtatgcgc</td>\n",
       "      <td>SASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5098505</th>\n",
       "      <td>5098506</td>\n",
       "      <td>5098512</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>gcgcgc</td>\n",
       "      <td>SASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5099288</th>\n",
       "      <td>5099289</td>\n",
       "      <td>5099297</td>\n",
       "      <td>908.3955</td>\n",
       "      <td>gtgcgcgt</td>\n",
       "      <td>SASASASA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29749 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Start      End   ZH-Score      Sequence  Conformation\n",
       "733          734      742   766.6232      cgtgcgtg      ASASASAS\n",
       "1677        1678     1690  1356.2120  gggcgtgcccgc  SASASASASASA\n",
       "1679        1680     1690  1201.6710    gcgtgcccgc    SASASASASA\n",
       "1744        1745     1753   516.0707      ttgcgcgc      SASASASA\n",
       "1746        1747     1753   883.5764        gcgcgc        SASASA\n",
       "...          ...      ...        ...           ...           ...\n",
       "5096951  5096952  5096962  2183.5740    cgtgcgtgtg    ASASASASAS\n",
       "5097858  5097859  5097871   717.9641  aagcgtatgcgc  SASASASASASA\n",
       "5097860  5097861  5097871   948.8341    gcgtatgcgc    SASASASASA\n",
       "5098505  5098506  5098512   883.5764        gcgcgc        SASASA\n",
       "5099288  5099289  5099297   908.3955      gtgcgcgt      SASASASA\n",
       "\n",
       "[29749 rows x 5 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhunt_dict[genomes_list[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_zhunt_data(df):\n",
    "    df = df.sort_values([\"End\", \"Start\"], ascending=True)\n",
    "\n",
    "    # Count length with overlapping positions\n",
    "    pos = 0\n",
    "    length = 0\n",
    "    for i, row in df.iterrows():\n",
    "        if row.Start > pos:\n",
    "            length += row.End - row.Start + 1\n",
    "            pos = row.End\n",
    "        elif row.Start < pos and row.End > pos:\n",
    "            length += row.End - pos + 1\n",
    "            pos = row.End\n",
    "    \n",
    "    return [len(df), length]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_billingiae</th>\n",
       "      <th>E_gerundensis</th>\n",
       "      <th>E_persicina</th>\n",
       "      <th>E_pyrifoliae</th>\n",
       "      <th>E_tasmaniensis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Number of Z-DNA regions (with overlaps)</th>\n",
       "      <td>29749</td>\n",
       "      <td>33366</td>\n",
       "      <td>25907</td>\n",
       "      <td>25346</td>\n",
       "      <td>26301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z-DNA length</th>\n",
       "      <td>119602</td>\n",
       "      <td>131756</td>\n",
       "      <td>107026</td>\n",
       "      <td>102084</td>\n",
       "      <td>103424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         E_billingiae  E_gerundensis  \\\n",
       "Number of Z-DNA regions (with overlaps)         29749          33366   \n",
       "Z-DNA length                                   119602         131756   \n",
       "\n",
       "                                         E_persicina  E_pyrifoliae  \\\n",
       "Number of Z-DNA regions (with overlaps)        25907         25346   \n",
       "Z-DNA length                                  107026        102084   \n",
       "\n",
       "                                         E_tasmaniensis  \n",
       "Number of Z-DNA regions (with overlaps)           26301  \n",
       "Z-DNA length                                     103424  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhunt_stats = dict()\n",
    "for genome in genomes_list:\n",
    "    zhunt_stats[genome] = process_zhunt_data(zhunt_dict[genome])\n",
    "\n",
    "pd.DataFrame(zhunt_stats, index=[\"Number of Z-DNA regions (with overlaps)\", \"Z-DNA length\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Рисуем гистограммы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "for genome in genomes_list:\n",
    "    sns.histplot(data=zhunt_dict[genome], x=\"ZH-Score\", bins=30, log_scale=True)\n",
    "    plt.title(genome.replace('_', '. ')+\" ZH-Scores\")\n",
    "    plt.savefig(f\"histograms/{genome}.png\", dpi=400)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Сохраняем данные и создаём .bed файлы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    zhunt_dict[genome].to_csv(f\"zhunt_result/{genome}.zhunt.txt\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    df = zhunt_dict[genomes_list[0]].copy()\n",
    "    bed_df = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": [genome] * len(df),\n",
    "            \"chromStart\": df.Start,\n",
    "            \"chromEnd\": df.End,\n",
    "            \"name\": df.index,\n",
    "            \"score\": df[\"ZH-Score\"]\n",
    "        }\n",
    "    )\n",
    "\n",
    "    bed_df.to_csv(f\"zhunt_result/{genome}.zhunt.bed\", index=None, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Объединяем пересечения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools merge -c 5 -o max -i zhunt_result/{genome}.zhunt.bed > zhunt_result/{genome}.zhunt.merged.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ассоциируем предсказанные участки Z-DNA с промоторами генов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "features_dict = dict()\n",
    "\n",
    "for genome in genomes_list:\n",
    "    features = pd.read_csv(f\"genomes/{genome}_feature_table.txt\", sep='\\t')\n",
    "    features = features.loc[\n",
    "        (features['seq_type']=='chromosome')\n",
    "    ]\n",
    "    locuses = list(np.unique(features.locus_tag))\n",
    "    df = features.copy()\n",
    "\n",
    "    counter = 1\n",
    "    for locus in locuses:\n",
    "        prod = list(features['product_accession'].loc[features.locus_tag == locus].dropna())\n",
    "        if prod == []:\n",
    "            prod = [\"RNA\" + str(counter)]\n",
    "            counter += 1\n",
    "            df['product_accession'].loc[df.locus_tag == locus] = prod[0]\n",
    "        elif len(prod) == 1:\n",
    "            df['product_accession'].loc[df.locus_tag == locus] = prod[0]\n",
    "        else:\n",
    "            print(prod)\n",
    "    df = df.loc[df['# feature'] == 'gene']\n",
    "    df = df[[\"# feature\", \"class\" , \"start\", \"end\", \"strand\", \"product_accession\", \"locus_tag\"]]\n",
    "    features_dict[genome] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    features_dict[genome]['TSS'] = features_dict[genome].apply(lambda x: x.start if x.strand == \"+\" else x.end, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    df = features_dict[genome].copy()\n",
    "    bed_df = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": [genome] * len(df),\n",
    "            \"chromStart\": df.TSS,\n",
    "            \"chromEnd\": df.TSS,\n",
    "            \"name\": df.product_accession\n",
    "        }\n",
    "    )\n",
    "\n",
    "    bed_df.to_csv(f\"TSS/{genome}.bed\", index=None, header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(\n",
    "    {\n",
    "        \"Species\": genomes_list,\n",
    "        \"Len\": [annot_dict[genome][2] for genome in genomes_list]\n",
    "    }\n",
    ").to_csv(f\"TSS/my.genomes\", index=None, header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools slop -g TSS/my.genomes -i TSS/{genome}.bed -b 100 > TSS/{genome}.slop.bed\")\n",
    "    os.system(f\"bedtools intersect -a TSS/{genome}.slop.bed -b zhunt_result/{genome}.zhunt.merged.bed -wb > intersect/{genome}.inter.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Создаём кластеры"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sudo apt install proteinortho"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****************************************************************\n",
      "\u001b[1;32mProteinortho\u001b[0m with PoFF version 6.0.14 - An orthology detection tool\n",
      "*****************************************************************\n",
      "Detected 8 available CPU threads (adjust this with -cpus), Detected 'diamond' version 0.9.30\n",
      "Checking input files.\n",
      "Checking genomes/E_billingiae_protein.faa... ok\n",
      "Checking genomes/E_gerundensis_protein.faa... ok\n",
      "Checking genomes/E_persicina_protein.faa... ok\n",
      "Checking genomes/E_pyrifoliae_protein.faa... ok\n",
      "Checking genomes/E_tasmaniensis_protein.faa... ok\n",
      "\n",
      "\u001b[1;32m**Step 1**\u001b[0m\n",
      "Generating indices.\n",
      "The database for 'genomes/E_billingiae_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_persicina_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_gerundensis_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_tasmaniensis_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_pyrifoliae_protein.faa' is present and will be used\n",
      "\n",
      "\u001b[1;32m**Step 2**\u001b[0m using diamond \n",
      "Running blast analysis: 100% (10/10)                                           \n",
      "[OUTPUT] -> written to erwinia.blast-graph\n",
      "\n",
      "\u001b[1;32m**Step 3**\u001b[0m\n",
      "Clustering by similarity (Proteinortho mode) using up to 5883 MB of memory (75% of total memory) and 8 cpu core(s). Adjust this behaviour with the -mem option.\n",
      "Reading erwinia.blast-graph\n",
      "5 species\n",
      "16925 paired proteins\n",
      "27875 bidirectional edges\n",
      "Done                       \n",
      "[OUTPUT] -> Orthologous groups are written to erwinia.proteinortho.tsv\n",
      "You can extract the fasta files of each orthology group with 'proteinortho_grab_proteins.pl -tofiles erwinia.proteinortho.tsv genomes/E_billingiae_protein.faa genomes/E_persicina_protein.faa genomes/E_gerundensis_protein.faa genomes/E_tasmaniensis_protein.faa genomes/E_pyrifoliae_protein.faa'\n",
      " (Careful: This will generate a file foreach line in the file erwinia.proteinortho.tsv).\n",
      "[OUTPUT] -> Orthologous pairs are written to erwinia.proteinortho-graph\n",
      "[OUTPUT] -> Summary is written to erwinia.proteinortho-graph.summary\n",
      "[OUTPUT] -> Orthologous groups are written to erwinia.proteinortho.html\n",
      "\n",
      "\u001b[1;32mAll finished.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!proteinortho -project=erwinia genomes/*faa\n",
    "!mv erwinia* ortho/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Пересекаем кластеры и Z-DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.982</td>\n",
       "      <td>WP_013204225.1</td>\n",
       "      <td>WP_067427185.1</td>\n",
       "      <td>WP_222681877.1</td>\n",
       "      <td>WP_015899125.1</td>\n",
       "      <td>WP_012442922.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_173363027.1</td>\n",
       "      <td>WP_067429596.1</td>\n",
       "      <td>WP_174522410.1</td>\n",
       "      <td>WP_104945051.1</td>\n",
       "      <td>WP_012441756.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013201311.1</td>\n",
       "      <td>WP_225700887.1</td>\n",
       "      <td>WP_062748436.1</td>\n",
       "      <td>WP_012668725.1</td>\n",
       "      <td>WP_012442032.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013200102.1</td>\n",
       "      <td>WP_067426621.1</td>\n",
       "      <td>WP_062746920.1</td>\n",
       "      <td>WP_012666468.1</td>\n",
       "      <td>WP_012439841.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013203471.1</td>\n",
       "      <td>WP_067433095.1</td>\n",
       "      <td>WP_245006712.1</td>\n",
       "      <td>WP_012667256.1</td>\n",
       "      <td>WP_193345526.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          5      5       0.982           WP_013204225.1   \n",
       "1          5      5       0.981           WP_173363027.1   \n",
       "2          5      5       0.981           WP_013201311.1   \n",
       "3          5      5       0.981           WP_013200102.1   \n",
       "4          5      5       0.981           WP_013203471.1   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0            WP_067427185.1          WP_222681877.1           WP_015899125.1   \n",
       "1            WP_067429596.1          WP_174522410.1           WP_104945051.1   \n",
       "2            WP_225700887.1          WP_062748436.1           WP_012668725.1   \n",
       "3            WP_067426621.1          WP_062746920.1           WP_012666468.1   \n",
       "4            WP_067433095.1          WP_245006712.1           WP_012667256.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  \n",
       "0             WP_012442922.1  \n",
       "1             WP_012441756.1  \n",
       "2             WP_012442032.1  \n",
       "3             WP_012439841.1  \n",
       "4             WP_193345526.1  "
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters = pd.read_csv(\"ortho/erwinia.proteinortho.tsv\", sep='\\t')\n",
    "clusters = clusters.loc[clusters['# Species']==5]\n",
    "clusters.sort_values(\"Alg.-Conn.\", ascending=False, inplace=True)\n",
    "clusters.reset_index(drop=True, inplace=True)\n",
    "clusters.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_zdna_df(genome):\n",
    "    zdna = pd.read_csv(f\"intersect/{genome}.inter.bed\", sep='\\t', \n",
    "                names=[\"Sp1\", \"St1\", \"En1\", \"Protein\", \"Sp2\", \"St2\", \"En2\", f\"Score_{genome}\"]\n",
    "                )[[\"Protein\", f\"Score_{genome}\"]]\n",
    "    zdna.sort_values(f\"Score_{genome}\", ascending=False, inplace=True)\n",
    "    # Drop duplicate proteins, keep maximal score\n",
    "    zdna.drop_duplicates(subset='Protein', inplace=True)\n",
    "    zdna.set_index(\"Protein\", inplace=True)\n",
    "    return zdna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.9820</td>\n",
       "      <td>WP_013204225.1</td>\n",
       "      <td>WP_067427185.1</td>\n",
       "      <td>WP_222681877.1</td>\n",
       "      <td>WP_015899125.1</td>\n",
       "      <td>WP_012442922.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.9810</td>\n",
       "      <td>WP_173363027.1</td>\n",
       "      <td>WP_067429596.1</td>\n",
       "      <td>WP_174522410.1</td>\n",
       "      <td>WP_104945051.1</td>\n",
       "      <td>WP_012441756.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.9810</td>\n",
       "      <td>WP_013201311.1</td>\n",
       "      <td>WP_225700887.1</td>\n",
       "      <td>WP_062748436.1</td>\n",
       "      <td>WP_012668725.1</td>\n",
       "      <td>WP_012442032.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.9810</td>\n",
       "      <td>WP_013200102.1</td>\n",
       "      <td>WP_067426621.1</td>\n",
       "      <td>WP_062746920.1</td>\n",
       "      <td>WP_012666468.1</td>\n",
       "      <td>WP_012439841.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.9810</td>\n",
       "      <td>WP_013203471.1</td>\n",
       "      <td>WP_067433095.1</td>\n",
       "      <td>WP_245006712.1</td>\n",
       "      <td>WP_012667256.1</td>\n",
       "      <td>WP_193345526.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2236</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0432</td>\n",
       "      <td>WP_013200514.1</td>\n",
       "      <td>WP_067433914.1</td>\n",
       "      <td>WP_245006688.1</td>\n",
       "      <td>WP_012667009.1</td>\n",
       "      <td>WP_012440387.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2237</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0362</td>\n",
       "      <td>WP_013199968.1</td>\n",
       "      <td>WP_067437212.1</td>\n",
       "      <td>WP_062749063.1</td>\n",
       "      <td>WP_011078071.1</td>\n",
       "      <td>WP_012443247.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2238</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0301</td>\n",
       "      <td>WP_013201440.1</td>\n",
       "      <td>WP_225701828.1</td>\n",
       "      <td>WP_118664384.1</td>\n",
       "      <td>WP_012668633.1</td>\n",
       "      <td>WP_012441938.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2239</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0265</td>\n",
       "      <td>WP_013202206.1</td>\n",
       "      <td>WP_225702167.1</td>\n",
       "      <td>WP_222681770.1</td>\n",
       "      <td>WP_014539578.1</td>\n",
       "      <td>WP_012442779.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2240</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0198</td>\n",
       "      <td>WP_013201360.1</td>\n",
       "      <td>WP_187486131.1</td>\n",
       "      <td>WP_187509559.1</td>\n",
       "      <td>WP_015899051.1</td>\n",
       "      <td>WP_012442741.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2241 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0             5      5      0.9820           WP_013204225.1   \n",
       "1             5      5      0.9810           WP_173363027.1   \n",
       "2             5      5      0.9810           WP_013201311.1   \n",
       "3             5      5      0.9810           WP_013200102.1   \n",
       "4             5      5      0.9810           WP_013203471.1   \n",
       "...         ...    ...         ...                      ...   \n",
       "2236          5      5      0.0432           WP_013200514.1   \n",
       "2237          5      5      0.0362           WP_013199968.1   \n",
       "2238          5      5      0.0301           WP_013201440.1   \n",
       "2239          5      5      0.0265           WP_013202206.1   \n",
       "2240          5      5      0.0198           WP_013201360.1   \n",
       "\n",
       "     E_gerundensis_protein.faa E_persicina_protein.faa  \\\n",
       "0               WP_067427185.1          WP_222681877.1   \n",
       "1               WP_067429596.1          WP_174522410.1   \n",
       "2               WP_225700887.1          WP_062748436.1   \n",
       "3               WP_067426621.1          WP_062746920.1   \n",
       "4               WP_067433095.1          WP_245006712.1   \n",
       "...                        ...                     ...   \n",
       "2236            WP_067433914.1          WP_245006688.1   \n",
       "2237            WP_067437212.1          WP_062749063.1   \n",
       "2238            WP_225701828.1          WP_118664384.1   \n",
       "2239            WP_225702167.1          WP_222681770.1   \n",
       "2240            WP_187486131.1          WP_187509559.1   \n",
       "\n",
       "     E_pyrifoliae_protein.faa E_tasmaniensis_protein.faa  \n",
       "0              WP_015899125.1             WP_012442922.1  \n",
       "1              WP_104945051.1             WP_012441756.1  \n",
       "2              WP_012668725.1             WP_012442032.1  \n",
       "3              WP_012666468.1             WP_012439841.1  \n",
       "4              WP_012667256.1             WP_193345526.1  \n",
       "...                       ...                        ...  \n",
       "2236           WP_012667009.1             WP_012440387.1  \n",
       "2237           WP_011078071.1             WP_012443247.1  \n",
       "2238           WP_012668633.1             WP_012441938.1  \n",
       "2239           WP_014539578.1             WP_012442779.1  \n",
       "2240           WP_015899051.1             WP_012442741.1  \n",
       "\n",
       "[2241 rows x 8 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "for genome in genomes_list:\n",
    "    clusters[\"E_billingiae_protein.faa\"].apply(lambda x: )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Score_E_gerundensis</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Protein</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WP_225700782.1</th>\n",
       "      <td>941334.2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_225700736.1</th>\n",
       "      <td>302785.5000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_187486806.1</th>\n",
       "      <td>138924.1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_067432241.1</th>\n",
       "      <td>138924.1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA73</th>\n",
       "      <td>138924.1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_225701014.1</th>\n",
       "      <td>505.8732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_187486789.1</th>\n",
       "      <td>505.8732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_225701050.1</th>\n",
       "      <td>505.8732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_067428513.1</th>\n",
       "      <td>505.8732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WP_067432615.1</th>\n",
       "      <td>500.8571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1088 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Score_E_gerundensis\n",
       "Protein                            \n",
       "WP_225700782.1          941334.2000\n",
       "WP_225700736.1          302785.5000\n",
       "WP_187486806.1          138924.1000\n",
       "WP_067432241.1          138924.1000\n",
       "RNA73                   138924.1000\n",
       "...                             ...\n",
       "WP_225701014.1             505.8732\n",
       "WP_187486789.1             505.8732\n",
       "WP_225701050.1             505.8732\n",
       "WP_067428513.1             505.8732\n",
       "WP_067432615.1             500.8571\n",
       "\n",
       "[1088 rows x 1 columns]"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_zdna_df(genomes_list[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
