{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Обязательная часть"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Скачивание файлов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! sudo apt install clustalw bedtools proteinortho"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "base = \"http://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/\"\n",
    "paths = {\n",
    "    \"E_billingiae\": \"000/196/615/GCF_000196615.1_ASM19661v1\",\n",
    "    \"E_gerundensis\": \"020/342/335/GCF_020342335.1_ASM2034233v1\",\n",
    "    \"E_persicina\": \"019/844/095/GCF_019844095.1_ASM1984409v1\",\n",
    "    \"E_pyrifoliae\": \"002/952/315/GCF_002952315.1_ASM295231v1\",\n",
    "    \"E_tasmaniensis\": \"000/026/185/GCF_000026185.1_ASM2618v1\"\n",
    "}\n",
    "genomic_end = \"_genomic.fna.gz\"\n",
    "feature_end = \"_feature_table.txt.gz\"\n",
    "protein_end = \"_protein.faa.gz\"\n",
    "\n",
    "! mkdir -p genomes histograms zhunt_result TSS intersect ortho clusters/raw clusters/aligned pictures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in paths.keys():\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{genomic_end} -O 'genomes/{key}{genomic_end}'\")\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{feature_end} -O 'genomes/{key}{feature_end}'\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{genomic_end}\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{feature_end}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in paths.keys():\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{protein_end} -O 'genomes/{key}{protein_end}'\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{protein_end}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Анализ аннотированных геномов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_list = list(paths.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annot_analyse(genome):\n",
    "    features = pd.read_csv(f\"genomes/{genome}_feature_table.txt\", sep='\\t')\n",
    "    features = features.loc[\n",
    "        (features['# feature']=='gene') & \n",
    "        (features['seq_type']=='chromosome')\n",
    "    ]\n",
    "    seqs = [seq_record for seq_record in SeqIO.parse(f\"genomes/{genome}_genomic.fna\", \"fasta\")]\n",
    "    # Choose complete genome, remove plasmids\n",
    "\n",
    "    genome_len = max(list(map(lambda x: len(x.seq), seqs)))\n",
    "    total_len = sum(list(map(lambda x: len(x.seq), seqs)))\n",
    "\n",
    "    total_genes_len = features[\"feature_interval_length\"].sum()\n",
    "\n",
    "    return [len(features), round(total_genes_len/genome_len*100, 2), genome_len, total_len, len(seqs)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_billingiae</th>\n",
       "      <th>E_gerundensis</th>\n",
       "      <th>E_persicina</th>\n",
       "      <th>E_pyrifoliae</th>\n",
       "      <th>E_tasmaniensis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Genes number</th>\n",
       "      <td>4703.00</td>\n",
       "      <td>3458.00</td>\n",
       "      <td>4311.00</td>\n",
       "      <td>3746.00</td>\n",
       "      <td>3569.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome coverage</th>\n",
       "      <td>88.99</td>\n",
       "      <td>88.19</td>\n",
       "      <td>88.61</td>\n",
       "      <td>86.27</td>\n",
       "      <td>86.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome length</th>\n",
       "      <td>5100167.00</td>\n",
       "      <td>3748909.00</td>\n",
       "      <td>4664605.00</td>\n",
       "      <td>4027225.00</td>\n",
       "      <td>3883467.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome + plasmid length</th>\n",
       "      <td>5372268.00</td>\n",
       "      <td>4437416.00</td>\n",
       "      <td>4802925.00</td>\n",
       "      <td>4075681.00</td>\n",
       "      <td>4067864.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Number of sequences</th>\n",
       "      <td>3.00</td>\n",
       "      <td>3.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>6.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         E_billingiae  E_gerundensis  E_persicina  \\\n",
       "Genes number                  4703.00        3458.00      4311.00   \n",
       "Genome coverage                 88.99          88.19        88.61   \n",
       "Genome length              5100167.00     3748909.00   4664605.00   \n",
       "Genome + plasmid length    5372268.00     4437416.00   4802925.00   \n",
       "Number of sequences              3.00           3.00         2.00   \n",
       "\n",
       "                         E_pyrifoliae  E_tasmaniensis  \n",
       "Genes number                  3746.00         3569.00  \n",
       "Genome coverage                 86.27           86.71  \n",
       "Genome length              4027225.00      3883467.00  \n",
       "Genome + plasmid length    4075681.00      4067864.00  \n",
       "Number of sequences              2.00            6.00  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    annot_dict[genome] = annot_analyse(genome)\n",
    "\n",
    "pd.DataFrame(annot_dict, \n",
    "index=[\"Genes number\", \"Genome coverage\", \n",
    "\"Genome length\", \"Genome + plasmid length\", \"Number of sequences\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Предсказание участков Z-DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system(\"wget -c -nv https://raw.githubusercontent.com/vanya-antonov/hse22-project/main/zhunt3-alan.c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system(\"gcc zhunt3-alan.c -lm -o zhunt3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from subprocess import DEVNULL, PIPE\n",
    "\n",
    "ZH_EXECUTABLE = Path(\"../hse22_project/zhunt3\")\n",
    "assert ZH_EXECUTABLE.is_file()\n",
    "\n",
    "def zhunt(query: str, windowsize: int = 6, minsize: int = 3, maxsize: int = 6):\n",
    "    assert set(query).issubset({\"A\", \"C\", \"G\", \"T\", \"N\"})\n",
    "    fd, temp = tempfile.mkstemp()\n",
    "    os.close(fd)\n",
    "    with open(temp, 'w') as stream:\n",
    "        stream.write(query)\n",
    "\n",
    "    subprocess.run(\n",
    "        [ZH_EXECUTABLE, \n",
    "         str(windowsize), str(minsize), str(maxsize), temp],\n",
    "        check=True, stdout=PIPE, stderr=DEVNULL,\n",
    "        input=query, encoding='ascii'\n",
    "    )\n",
    "    with open(temp + \".Z-SCORE\", 'r') as stream:\n",
    "        df = pd.read_csv(stream,\n",
    "                         names=['Start', 'End', 'nu-1', 'nu-2', 'nu-3', \n",
    "                                'ZH-Score', 'Sequence', 'Conformation'],\n",
    "                         skiprows=1, sep='\\s+')\n",
    "    os.remove(temp)\n",
    "    os.remove(temp + \".Z-SCORE\")\n",
    "    df = df.loc[df[\"ZH-Score\"] > 500]\n",
    "    return df[['Start', 'End', 'ZH-Score', 'Sequence', 'Conformation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [],
   "source": [
    "def choose_max_seq(genome):\n",
    "    seqs = [str(seq_record.seq) for seq_record in SeqIO.parse(f\"genomes/{genome}_genomic.fna\", \"fasta\")]\n",
    "    # Choose complete genome, remove plasmids\n",
    "    max_scaffold = max(seqs, key=len)\n",
    "    return max_scaffold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [],
   "source": [
    "zhunt_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    zhunt_dict[genome] = zhunt(choose_max_seq(genome))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ZH-Score</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Conformation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>734</td>\n",
       "      <td>742</td>\n",
       "      <td>766.6232</td>\n",
       "      <td>cgtgcgtg</td>\n",
       "      <td>ASASASAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1677</th>\n",
       "      <td>1678</td>\n",
       "      <td>1690</td>\n",
       "      <td>1356.2120</td>\n",
       "      <td>gggcgtgcccgc</td>\n",
       "      <td>SASASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1679</th>\n",
       "      <td>1680</td>\n",
       "      <td>1690</td>\n",
       "      <td>1201.6710</td>\n",
       "      <td>gcgtgcccgc</td>\n",
       "      <td>SASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>1745</td>\n",
       "      <td>1753</td>\n",
       "      <td>516.0707</td>\n",
       "      <td>ttgcgcgc</td>\n",
       "      <td>SASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1746</th>\n",
       "      <td>1747</td>\n",
       "      <td>1753</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>gcgcgc</td>\n",
       "      <td>SASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5096951</th>\n",
       "      <td>5096952</td>\n",
       "      <td>5096962</td>\n",
       "      <td>2183.5740</td>\n",
       "      <td>cgtgcgtgtg</td>\n",
       "      <td>ASASASASAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5097858</th>\n",
       "      <td>5097859</td>\n",
       "      <td>5097871</td>\n",
       "      <td>717.9641</td>\n",
       "      <td>aagcgtatgcgc</td>\n",
       "      <td>SASASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5097860</th>\n",
       "      <td>5097861</td>\n",
       "      <td>5097871</td>\n",
       "      <td>948.8341</td>\n",
       "      <td>gcgtatgcgc</td>\n",
       "      <td>SASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5098505</th>\n",
       "      <td>5098506</td>\n",
       "      <td>5098512</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>gcgcgc</td>\n",
       "      <td>SASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5099288</th>\n",
       "      <td>5099289</td>\n",
       "      <td>5099297</td>\n",
       "      <td>908.3955</td>\n",
       "      <td>gtgcgcgt</td>\n",
       "      <td>SASASASA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29749 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Start      End   ZH-Score      Sequence  Conformation\n",
       "733          734      742   766.6232      cgtgcgtg      ASASASAS\n",
       "1677        1678     1690  1356.2120  gggcgtgcccgc  SASASASASASA\n",
       "1679        1680     1690  1201.6710    gcgtgcccgc    SASASASASA\n",
       "1744        1745     1753   516.0707      ttgcgcgc      SASASASA\n",
       "1746        1747     1753   883.5764        gcgcgc        SASASA\n",
       "...          ...      ...        ...           ...           ...\n",
       "5096951  5096952  5096962  2183.5740    cgtgcgtgtg    ASASASASAS\n",
       "5097858  5097859  5097871   717.9641  aagcgtatgcgc  SASASASASASA\n",
       "5097860  5097861  5097871   948.8341    gcgtatgcgc    SASASASASA\n",
       "5098505  5098506  5098512   883.5764        gcgcgc        SASASA\n",
       "5099288  5099289  5099297   908.3955      gtgcgcgt      SASASASA\n",
       "\n",
       "[29749 rows x 5 columns]"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhunt_dict[genomes_list[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_zhunt_data(df):\n",
    "    df = df.sort_values([\"End\", \"Start\"], ascending=True)\n",
    "\n",
    "    # Count length with overlapping positions\n",
    "    pos = 0\n",
    "    length = 0\n",
    "    for i, row in df.iterrows():\n",
    "        if row.Start > pos:\n",
    "            length += row.End - row.Start + 1\n",
    "            pos = row.End\n",
    "        elif row.Start < pos and row.End > pos:\n",
    "            length += row.End - pos + 1\n",
    "            pos = row.End\n",
    "    \n",
    "    return [len(df), length]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_billingiae</th>\n",
       "      <th>E_gerundensis</th>\n",
       "      <th>E_persicina</th>\n",
       "      <th>E_pyrifoliae</th>\n",
       "      <th>E_tasmaniensis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Number of Z-DNA regions (with overlaps)</th>\n",
       "      <td>29749</td>\n",
       "      <td>33366</td>\n",
       "      <td>25907</td>\n",
       "      <td>25346</td>\n",
       "      <td>26301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z-DNA length</th>\n",
       "      <td>119602</td>\n",
       "      <td>131756</td>\n",
       "      <td>107026</td>\n",
       "      <td>102084</td>\n",
       "      <td>103424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         E_billingiae  E_gerundensis  \\\n",
       "Number of Z-DNA regions (with overlaps)         29749          33366   \n",
       "Z-DNA length                                   119602         131756   \n",
       "\n",
       "                                         E_persicina  E_pyrifoliae  \\\n",
       "Number of Z-DNA regions (with overlaps)        25907         25346   \n",
       "Z-DNA length                                  107026        102084   \n",
       "\n",
       "                                         E_tasmaniensis  \n",
       "Number of Z-DNA regions (with overlaps)           26301  \n",
       "Z-DNA length                                     103424  "
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhunt_stats = dict()\n",
    "for genome in genomes_list:\n",
    "    zhunt_stats[genome] = process_zhunt_data(zhunt_dict[genome])\n",
    "\n",
    "pd.DataFrame(zhunt_stats, index=[\"Number of Z-DNA regions (with overlaps)\", \"Z-DNA length\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Рисуем гистограммы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "for genome in genomes_list:\n",
    "    sns.histplot(data=zhunt_dict[genome], x=\"ZH-Score\", bins=30, log_scale=True)\n",
    "    plt.title(genome.replace('_', '. ')+\" ZH-Scores\")\n",
    "    plt.savefig(f\"histograms/{genome}.png\", dpi=400)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Сохраняем данные и создаём .bed файлы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    zhunt_dict[genome].to_csv(f\"zhunt_result/{genome}.zhunt.txt\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    df = zhunt_dict[genome].copy()\n",
    "    bed_df = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": [genome] * len(df),\n",
    "            \"chromStart\": df.Start,\n",
    "            \"chromEnd\": df.End,\n",
    "            \"name\": df.index,\n",
    "            \"score\": df[\"ZH-Score\"]\n",
    "        }\n",
    "    )\n",
    "\n",
    "    bed_df.to_csv(f\"zhunt_result/{genome}.zhunt.bed\", index=None, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Объединяем пересечения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools merge -c 5 -o max -i zhunt_result/{genome}.zhunt.bed > zhunt_result/{genome}.zhunt.merged.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ассоциируем предсказанные участки Z-DNA с промоторами генов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 369,
   "metadata": {},
   "outputs": [],
   "source": [
    "features_dict = dict()\n",
    "\n",
    "for genome in genomes_list:\n",
    "    features = pd.read_csv(f\"genomes/{genome}_feature_table.txt\", sep='\\t')\n",
    "    features = features.loc[\n",
    "        (features['seq_type']=='chromosome')\n",
    "    ]\n",
    "    locuses = list(np.unique(features.locus_tag))\n",
    "    df = features.copy()\n",
    "\n",
    "    counter = 1\n",
    "    for locus in locuses:\n",
    "        prod = list(features['product_accession'].loc[features.locus_tag == locus].dropna())\n",
    "        if prod == []:\n",
    "            prod = [\"RNA\" + str(counter)]\n",
    "            counter += 1\n",
    "            df['product_accession'].loc[df.locus_tag == locus] = prod[0]\n",
    "        elif len(prod) == 1:\n",
    "            df['product_accession'].loc[df.locus_tag == locus] = prod[0]\n",
    "        else:\n",
    "            print(prod)\n",
    "    df = df.loc[df['# feature'] == 'gene']\n",
    "    df = df[[\"# feature\", \"class\" , \"start\", \"end\", \"strand\", \"product_accession\", \"locus_tag\"]]\n",
    "    features_dict[genome] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    features_dict[genome]['TSS'] = features_dict[genome].apply(lambda x: x.start if x.strand == \"+\" else x.end, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    df = features_dict[genome].copy()\n",
    "    bed_df = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": [genome] * len(df),\n",
    "            \"chromStart\": df.TSS,\n",
    "            \"chromEnd\": df.TSS,\n",
    "            \"name\": df.product_accession\n",
    "        }\n",
    "    )\n",
    "\n",
    "    bed_df.to_csv(f\"TSS/{genome}.bed\", index=None, header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(\n",
    "    {\n",
    "        \"Species\": genomes_list,\n",
    "        \"Len\": [annot_dict[genome][2] for genome in genomes_list]\n",
    "    }\n",
    ").to_csv(f\"TSS/my.genomes\", index=None, header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools slop -g TSS/my.genomes -i TSS/{genome}.bed -b 100 > TSS/{genome}.slop.bed\")\n",
    "    os.system(f\"bedtools intersect -a TSS/{genome}.slop.bed -b zhunt_result/{genome}.zhunt.merged.bed -wb > intersect/{genome}.inter.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Создаём кластеры"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****************************************************************\n",
      "\u001b[1;32mProteinortho\u001b[0m with PoFF version 6.0.14 - An orthology detection tool\n",
      "*****************************************************************\n",
      "Detected 8 available CPU threads (adjust this with -cpus), Detected 'diamond' version 0.9.30\n",
      "Checking input files.\n",
      "Checking genomes/E_billingiae_protein.faa... ok\n",
      "Checking genomes/E_gerundensis_protein.faa... ok\n",
      "Checking genomes/E_persicina_protein.faa... ok\n",
      "Checking genomes/E_pyrifoliae_protein.faa... ok\n",
      "Checking genomes/E_tasmaniensis_protein.faa... ok\n",
      "\n",
      "\u001b[1;32m**Step 1**\u001b[0m\n",
      "Generating indices.\n",
      "The database for 'genomes/E_billingiae_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_persicina_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_gerundensis_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_tasmaniensis_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_pyrifoliae_protein.faa' is present and will be used\n",
      "\n",
      "\u001b[1;32m**Step 2**\u001b[0m using diamond \n",
      "Running blast analysis: 100% (10/10)                                           \n",
      "[OUTPUT] -> written to erwinia.blast-graph\n",
      "\n",
      "\u001b[1;32m**Step 3**\u001b[0m\n",
      "Clustering by similarity (Proteinortho mode) using up to 5883 MB of memory (75% of total memory) and 8 cpu core(s). Adjust this behaviour with the -mem option.\n",
      "Reading erwinia.blast-graph\n",
      "5 species\n",
      "16925 paired proteins\n",
      "27875 bidirectional edges\n",
      "Done                       \n",
      "[OUTPUT] -> Orthologous groups are written to erwinia.proteinortho.tsv\n",
      "You can extract the fasta files of each orthology group with 'proteinortho_grab_proteins.pl -tofiles erwinia.proteinortho.tsv genomes/E_billingiae_protein.faa genomes/E_persicina_protein.faa genomes/E_gerundensis_protein.faa genomes/E_tasmaniensis_protein.faa genomes/E_pyrifoliae_protein.faa'\n",
      " (Careful: This will generate a file foreach line in the file erwinia.proteinortho.tsv).\n",
      "[OUTPUT] -> Orthologous pairs are written to erwinia.proteinortho-graph\n",
      "[OUTPUT] -> Summary is written to erwinia.proteinortho-graph.summary\n",
      "[OUTPUT] -> Orthologous groups are written to erwinia.proteinortho.html\n",
      "\n",
      "\u001b[1;32mAll finished.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!proteinortho -project=erwinia genomes/*faa\n",
    "!mv erwinia* ortho/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Статистика по кластерам"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WP_013203652.1</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012440702.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669311.1</td>\n",
       "      <td>WP_012440279.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669308.1</td>\n",
       "      <td>WP_012440282.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669307.1</td>\n",
       "      <td>WP_012440283.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669404.1</td>\n",
       "      <td>WP_012440317.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          2      2         1.0           WP_013203652.1   \n",
       "1          2      2         1.0                        *   \n",
       "2          2      2         1.0                        *   \n",
       "3          2      2         1.0                        *   \n",
       "4          2      2         1.0                        *   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0                         *                       *                        *   \n",
       "1                         *                       *           WP_012669311.1   \n",
       "2                         *                       *           WP_012669308.1   \n",
       "3                         *                       *           WP_012669307.1   \n",
       "4                         *                       *           WP_012669404.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  \n",
       "0             WP_012440702.1  \n",
       "1             WP_012440279.1  \n",
       "2             WP_012440282.1  \n",
       "3             WP_012440283.1  \n",
       "4             WP_012440317.1  "
      ]
     },
     "execution_count": 371,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters = pd.read_csv(\"ortho/erwinia.proteinortho.tsv\", sep='\\t')\n",
    "clusters.sort_values(\"Alg.-Conn.\", ascending=False, inplace=True)\n",
    "clusters.reset_index(drop=True, inplace=True)\n",
    "clusters.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4337"
      ]
     },
     "execution_count": 372,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_of_clusters = len(clusters)\n",
    "number_of_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "sns.countplot(data=clusters, x=\"# Species\")\n",
    "plt.title(\"Distribution of species in clusters\")\n",
    "#plt.xticks(ticks=[2, 3, 4, 5], labels=['2', '3', '4', '5'])\n",
    "plt.tight_layout()\n",
    "plt.savefig(f\"histograms/clusters.png\", dpi=400)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Пересекаем кластеры и Z-DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.982</td>\n",
       "      <td>WP_013204225.1</td>\n",
       "      <td>WP_067427185.1</td>\n",
       "      <td>WP_222681877.1</td>\n",
       "      <td>WP_015899125.1</td>\n",
       "      <td>WP_012442922.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_173363027.1</td>\n",
       "      <td>WP_067429596.1</td>\n",
       "      <td>WP_174522410.1</td>\n",
       "      <td>WP_104945051.1</td>\n",
       "      <td>WP_012441756.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013201311.1</td>\n",
       "      <td>WP_225700887.1</td>\n",
       "      <td>WP_062748436.1</td>\n",
       "      <td>WP_012668725.1</td>\n",
       "      <td>WP_012442032.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013200102.1</td>\n",
       "      <td>WP_067426621.1</td>\n",
       "      <td>WP_062746920.1</td>\n",
       "      <td>WP_012666468.1</td>\n",
       "      <td>WP_012439841.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013203471.1</td>\n",
       "      <td>WP_067433095.1</td>\n",
       "      <td>WP_245006712.1</td>\n",
       "      <td>WP_012667256.1</td>\n",
       "      <td>WP_193345526.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          5      5       0.982           WP_013204225.1   \n",
       "1          5      5       0.981           WP_173363027.1   \n",
       "2          5      5       0.981           WP_013201311.1   \n",
       "3          5      5       0.981           WP_013200102.1   \n",
       "4          5      5       0.981           WP_013203471.1   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0            WP_067427185.1          WP_222681877.1           WP_015899125.1   \n",
       "1            WP_067429596.1          WP_174522410.1           WP_104945051.1   \n",
       "2            WP_225700887.1          WP_062748436.1           WP_012668725.1   \n",
       "3            WP_067426621.1          WP_062746920.1           WP_012666468.1   \n",
       "4            WP_067433095.1          WP_245006712.1           WP_012667256.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  \n",
       "0             WP_012442922.1  \n",
       "1             WP_012441756.1  \n",
       "2             WP_012442032.1  \n",
       "3             WP_012439841.1  \n",
       "4             WP_193345526.1  "
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters = pd.read_csv(\"ortho/erwinia.proteinortho.tsv\", sep='\\t')\n",
    "clusters = clusters.loc[clusters['# Species']==5]\n",
    "clusters.sort_values(\"Alg.-Conn.\", ascending=False, inplace=True)\n",
    "clusters.reset_index(drop=True, inplace=True)\n",
    "clusters.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_zdna_df(genome):\n",
    "    zdna = pd.read_csv(f\"intersect/{genome}.inter.bed\", sep='\\t', \n",
    "                names=[\"Sp1\", \"St1\", \"En1\", \"Protein\", \"Sp2\", \"Start\", \"End\", f\"Score_{genome}\"]\n",
    "                )[[\"Protein\", f\"Score_{genome}\", \"Start\", \"End\"]]\n",
    "    zdna.sort_values(f\"Score_{genome}\", ascending=False, inplace=True)\n",
    "    # Drop duplicate proteins, keep maximal score\n",
    "    zdna.drop_duplicates(subset='Protein', inplace=True)\n",
    "    zdna.set_index(\"Protein\", inplace=True)\n",
    "    return zdna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "      <th>E_billingiae_protein.score</th>\n",
       "      <th>E_gerundensis_protein.score</th>\n",
       "      <th>E_persicina_protein.score</th>\n",
       "      <th>E_pyrifoliae_protein.score</th>\n",
       "      <th>E_tasmaniensis_protein.score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.982</td>\n",
       "      <td>WP_013204225.1</td>\n",
       "      <td>WP_067427185.1</td>\n",
       "      <td>WP_222681877.1</td>\n",
       "      <td>WP_015899125.1</td>\n",
       "      <td>WP_012442922.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28780.500</td>\n",
       "      <td>766.6232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_173363027.1</td>\n",
       "      <td>WP_067429596.1</td>\n",
       "      <td>WP_174522410.1</td>\n",
       "      <td>WP_104945051.1</td>\n",
       "      <td>WP_012441756.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2943.461</td>\n",
       "      <td>2943.4610</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1790.574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013201311.1</td>\n",
       "      <td>WP_225700887.1</td>\n",
       "      <td>WP_062748436.1</td>\n",
       "      <td>WP_012668725.1</td>\n",
       "      <td>WP_012442032.1</td>\n",
       "      <td>908.3955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013200102.1</td>\n",
       "      <td>WP_067426621.1</td>\n",
       "      <td>WP_062746920.1</td>\n",
       "      <td>WP_012666468.1</td>\n",
       "      <td>WP_012439841.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1733.503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>783.823</td>\n",
       "      <td>865.283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013203471.1</td>\n",
       "      <td>WP_067433095.1</td>\n",
       "      <td>WP_245006712.1</td>\n",
       "      <td>WP_012667256.1</td>\n",
       "      <td>WP_193345526.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          5      5       0.982           WP_013204225.1   \n",
       "1          5      5       0.981           WP_173363027.1   \n",
       "2          5      5       0.981           WP_013201311.1   \n",
       "3          5      5       0.981           WP_013200102.1   \n",
       "4          5      5       0.981           WP_013203471.1   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0            WP_067427185.1          WP_222681877.1           WP_015899125.1   \n",
       "1            WP_067429596.1          WP_174522410.1           WP_104945051.1   \n",
       "2            WP_225700887.1          WP_062748436.1           WP_012668725.1   \n",
       "3            WP_067426621.1          WP_062746920.1           WP_012666468.1   \n",
       "4            WP_067433095.1          WP_245006712.1           WP_012667256.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  E_billingiae_protein.score  \\\n",
       "0             WP_012442922.1                         NaN   \n",
       "1             WP_012441756.1                         NaN   \n",
       "2             WP_012442032.1                    908.3955   \n",
       "3             WP_012439841.1                         NaN   \n",
       "4             WP_193345526.1                         NaN   \n",
       "\n",
       "   E_gerundensis_protein.score  E_persicina_protein.score  \\\n",
       "0                    28780.500                   766.6232   \n",
       "1                     2943.461                  2943.4610   \n",
       "2                          NaN                        NaN   \n",
       "3                     1733.503                        NaN   \n",
       "4                          NaN                        NaN   \n",
       "\n",
       "   E_pyrifoliae_protein.score  E_tasmaniensis_protein.score  \n",
       "0                         NaN                           NaN  \n",
       "1                         NaN                      1790.574  \n",
       "2                         NaN                           NaN  \n",
       "3                     783.823                       865.283  \n",
       "4                         NaN                           NaN  "
      ]
     },
     "execution_count": 375,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for genome in genomes_list:\n",
    "    zdna = read_zdna_df(genome)[f\"Score_{genome}\"]\n",
    "    list_of_zhscores = [float(zdna.loc[prot]) if prot in list(zdna.index) else None for prot in clusters[f\"{genome}_protein.faa\"]]\n",
    "    clusters[f\"{genome}_protein.score\"] = list_of_zhscores\n",
    "clusters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chosen = clusters.dropna(subset=clusters.columns[-5:])\n",
    "means = chosen.iloc[:, 8:13].mean(axis=1)\n",
    "chosen[\"mean_score\"] = means\n",
    "chosen.sort_values(\"mean_score\", ascending=False, inplace=True)\n",
    "chosen = chosen.iloc[:8,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "      <th>E_billingiae_protein.score</th>\n",
       "      <th>E_gerundensis_protein.score</th>\n",
       "      <th>E_persicina_protein.score</th>\n",
       "      <th>E_pyrifoliae_protein.score</th>\n",
       "      <th>E_tasmaniensis_protein.score</th>\n",
       "      <th>mean_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>927</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.936</td>\n",
       "      <td>WP_013201010.1</td>\n",
       "      <td>WP_067428338.1</td>\n",
       "      <td>WP_118664543.1</td>\n",
       "      <td>WP_012668911.1</td>\n",
       "      <td>WP_012442250.1</td>\n",
       "      <td>138924.1000</td>\n",
       "      <td>28780.500</td>\n",
       "      <td>13713.9900</td>\n",
       "      <td>138924.1000</td>\n",
       "      <td>138924.1000</td>\n",
       "      <td>91853.35800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.950</td>\n",
       "      <td>WP_013200749.1</td>\n",
       "      <td>WP_067427946.1</td>\n",
       "      <td>WP_062747519.1</td>\n",
       "      <td>WP_012667092.1</td>\n",
       "      <td>WP_012440496.1</td>\n",
       "      <td>3700.9720</td>\n",
       "      <td>65884.110</td>\n",
       "      <td>4876.7350</td>\n",
       "      <td>302785.5000</td>\n",
       "      <td>21732.3800</td>\n",
       "      <td>79795.93940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1306</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.922</td>\n",
       "      <td>WP_041691924.1</td>\n",
       "      <td>WP_225700936.1</td>\n",
       "      <td>WP_062747774.1</td>\n",
       "      <td>WP_014539290.1</td>\n",
       "      <td>WP_042959064.1</td>\n",
       "      <td>3428.5290</td>\n",
       "      <td>4263.555</td>\n",
       "      <td>66470.7800</td>\n",
       "      <td>198956.2000</td>\n",
       "      <td>27872.6600</td>\n",
       "      <td>60198.34480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.961</td>\n",
       "      <td>WP_013201926.1</td>\n",
       "      <td>WP_225700696.1</td>\n",
       "      <td>WP_062743475.1</td>\n",
       "      <td>WP_012668206.1</td>\n",
       "      <td>WP_012441548.1</td>\n",
       "      <td>826.8355</td>\n",
       "      <td>904.320</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>211093.9000</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>42918.44166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>701</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.944</td>\n",
       "      <td>WP_013204048.1</td>\n",
       "      <td>WP_209500727.1</td>\n",
       "      <td>WP_062744481.1</td>\n",
       "      <td>WP_012666707.1</td>\n",
       "      <td>WP_012440081.1</td>\n",
       "      <td>138924.1000</td>\n",
       "      <td>65884.110</td>\n",
       "      <td>3661.1100</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>42047.29456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.973</td>\n",
       "      <td>WP_013201017.1</td>\n",
       "      <td>WP_225700938.1</td>\n",
       "      <td>WP_118664541.1</td>\n",
       "      <td>WP_012668905.1</td>\n",
       "      <td>WP_012442243.1</td>\n",
       "      <td>13713.9900</td>\n",
       "      <td>3993.322</td>\n",
       "      <td>138924.1000</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>31679.71296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>268</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.959</td>\n",
       "      <td>WP_013200502.1</td>\n",
       "      <td>WP_067433954.1</td>\n",
       "      <td>WP_062745643.1</td>\n",
       "      <td>WP_004160101.1</td>\n",
       "      <td>WP_012442652.1</td>\n",
       "      <td>1090.5640</td>\n",
       "      <td>38833.580</td>\n",
       "      <td>38833.5800</td>\n",
       "      <td>38833.5800</td>\n",
       "      <td>38833.5800</td>\n",
       "      <td>31284.97680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.935</td>\n",
       "      <td>WP_013200413.1</td>\n",
       "      <td>WP_067434130.1</td>\n",
       "      <td>WP_062748062.1</td>\n",
       "      <td>WP_014539549.1</td>\n",
       "      <td>WP_012442757.1</td>\n",
       "      <td>28780.5000</td>\n",
       "      <td>27872.660</td>\n",
       "      <td>27872.6600</td>\n",
       "      <td>27872.6600</td>\n",
       "      <td>27872.6600</td>\n",
       "      <td>28054.22800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "927           5      5       0.936           WP_013201010.1   \n",
       "531           5      5       0.950           WP_013200749.1   \n",
       "1306          5      5       0.922           WP_041691924.1   \n",
       "230           5      5       0.961           WP_013201926.1   \n",
       "701           5      5       0.944           WP_013204048.1   \n",
       "41            5      5       0.973           WP_013201017.1   \n",
       "268           5      5       0.959           WP_013200502.1   \n",
       "974           5      5       0.935           WP_013200413.1   \n",
       "\n",
       "     E_gerundensis_protein.faa E_persicina_protein.faa  \\\n",
       "927             WP_067428338.1          WP_118664543.1   \n",
       "531             WP_067427946.1          WP_062747519.1   \n",
       "1306            WP_225700936.1          WP_062747774.1   \n",
       "230             WP_225700696.1          WP_062743475.1   \n",
       "701             WP_209500727.1          WP_062744481.1   \n",
       "41              WP_225700938.1          WP_118664541.1   \n",
       "268             WP_067433954.1          WP_062745643.1   \n",
       "974             WP_067434130.1          WP_062748062.1   \n",
       "\n",
       "     E_pyrifoliae_protein.faa E_tasmaniensis_protein.faa  \\\n",
       "927            WP_012668911.1             WP_012442250.1   \n",
       "531            WP_012667092.1             WP_012440496.1   \n",
       "1306           WP_014539290.1             WP_042959064.1   \n",
       "230            WP_012668206.1             WP_012441548.1   \n",
       "701            WP_012666707.1             WP_012440081.1   \n",
       "41             WP_012668905.1             WP_012442243.1   \n",
       "268            WP_004160101.1             WP_012442652.1   \n",
       "974            WP_014539549.1             WP_012442757.1   \n",
       "\n",
       "      E_billingiae_protein.score  E_gerundensis_protein.score  \\\n",
       "927                  138924.1000                    28780.500   \n",
       "531                    3700.9720                    65884.110   \n",
       "1306                   3428.5290                     4263.555   \n",
       "230                     826.8355                      904.320   \n",
       "701                  138924.1000                    65884.110   \n",
       "41                    13713.9900                     3993.322   \n",
       "268                    1090.5640                    38833.580   \n",
       "974                   28780.5000                    27872.660   \n",
       "\n",
       "      E_persicina_protein.score  E_pyrifoliae_protein.score  \\\n",
       "927                  13713.9900                 138924.1000   \n",
       "531                   4876.7350                 302785.5000   \n",
       "1306                 66470.7800                 198956.2000   \n",
       "230                    883.5764                 211093.9000   \n",
       "701                   3661.1100                    883.5764   \n",
       "41                  138924.1000                    883.5764   \n",
       "268                  38833.5800                  38833.5800   \n",
       "974                  27872.6600                  27872.6600   \n",
       "\n",
       "      E_tasmaniensis_protein.score   mean_score  \n",
       "927                    138924.1000  91853.35800  \n",
       "531                     21732.3800  79795.93940  \n",
       "1306                    27872.6600  60198.34480  \n",
       "230                       883.5764  42918.44166  \n",
       "701                       883.5764  42047.29456  \n",
       "41                        883.5764  31679.71296  \n",
       "268                     38833.5800  31284.97680  \n",
       "974                     27872.6600  28054.22800  "
      ]
     },
     "execution_count": 377,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Выравниваем белки"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_seqs = dict()\n",
    "for genome in genomes_list:\n",
    "    aa_seqs[genome] = SeqIO.to_dict(SeqIO.parse(f\"genomes/{genome}_protein.faa\", \"fasta\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 1\n",
    "\n",
    "for i, row in chosen.iterrows():\n",
    "    seqs_in_cluster = []\n",
    "    for genome in genomes_list:\n",
    "        protein_id = row[f\"{genome}_protein.faa\"]\n",
    "        seqs_in_cluster.append(aa_seqs[genome][protein_id])\n",
    "    SeqIO.write(seqs_in_cluster, f\"clusters/raw/cluster_{counter}.faa\", \"fasta\")\n",
    "    counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cl in range(1,9):\n",
    "    os.system(f\"clustalw -align -type=PROTEIN -infile='clusters/raw/cluster_{cl}.faa' -OUTFILE='clusters/aligned/cluster_{cl}.aln'\")\n",
    "! rm -rf clusters/raw/*.dnd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Визуализируем"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dna_features_viewer import GraphicFeature, GraphicRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "metadata": {},
   "outputs": [],
   "source": [
    "zdna_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    features_dict[genome].set_index(\"product_accession\", drop=False, inplace=True)\n",
    "    zdna_dict[genome] = read_zdna_df(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 1\n",
    "for k, row in chosen.iterrows():\n",
    "    fig, axes = plt.subplots(nrows=5, ncols=1,figsize=(12, 6))\n",
    "    fig.suptitle(f'Cluster {counter}')\n",
    "\n",
    "    for i, genome in enumerate(genomes_list):\n",
    "        zdna = zdna_dict[genome]\n",
    "        features = features_dict[genome]\n",
    "\n",
    "        prot = row.loc[genome + '_protein.faa']\n",
    "        Z_DNA_label = \"Z-DNA, ZH-Score=\" + str(round(zdna.loc[prot][f\"Score_{genome}\"]))\n",
    "        coords = [zdna.loc[prot, \"Start\"], zdna.loc[prot, \"End\"],\n",
    "                  features.loc[prot, \"start\"], features.loc[prot, \"end\"]]\n",
    "        minimum = min(coords)\n",
    "\n",
    "        strand = +1 if features.loc[prot, \"strand\"]==\"+\" else -1\n",
    "\n",
    "        features=[\n",
    "        GraphicFeature(start=coords[0]-minimum+26, end=coords[1]-minimum+26, \n",
    "                   strand=+1, color=\"#0057b8\", label=Z_DNA_label),\n",
    "        GraphicFeature(start=coords[2]-minimum+26, end=coords[3]-minimum+26, \n",
    "                   strand=strand, color=\"#ffd700\", label=genome + \": \" + prot),\n",
    "                   ]\n",
    "        record = GraphicRecord(sequence_length=max(coords)-minimum+50, features=features)\n",
    "        record.plot(ax=axes[i])\n",
    "    plt.savefig(f'pictures/Cluster_{counter}.png', dpi=500)\n",
    "    counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Бонусная часть"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Поиск G-квадруплексов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "na_seqs = dict()\n",
    "for genome in genomes_list:\n",
    "    na_seqs[genome] = choose_max_seq(genome)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ставим PQSfinder и запускаем"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! Rscript configure.r"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "В `pqsfinder.r` немного ослабили ограничения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from subprocess import DEVNULL, PIPE\n",
    "\n",
    "Q_EXECUTABLE = Path(\"pqsfinder.r\")\n",
    "assert Q_EXECUTABLE.is_file()\n",
    "\n",
    "\n",
    "def pqs_exec(query: str):\n",
    "    assert set(query).issubset({\"A\", \"C\", \"G\", \"T\", \"N\"})\n",
    "    fd, temp = tempfile.mkstemp()\n",
    "    os.close(fd)\n",
    "    with open(temp, 'w') as stream:\n",
    "        stream.write(\">chr\\n\"+query)\n",
    "\n",
    "    subprocess.run(\n",
    "        [\"Rscript\", Q_EXECUTABLE, temp, temp + \".qgrs\"],\n",
    "        check=True, stdout=PIPE, stderr=DEVNULL,\n",
    "        input=query, encoding='ascii'\n",
    "    )\n",
    "    \n",
    "    with open(temp + \".qgrs\", 'r') as stream:\n",
    "        df = pd.read_csv(stream,\n",
    "                         #names=['ID', 'T1', 'T2', 'T3', 'T4', 'TS', 'GS', 'SEQ'],\n",
    "                         skiprows=0, sep=' ').iloc[:,:8]\n",
    "    os.remove(temp)\n",
    "    os.remove(temp + \".qgrs\")\n",
    "    #df = df.loc[df[\"score\"] > 500]\n",
    "    #return df\n",
    "    return df[['start', 'end', 'score', 'strand']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [],
   "source": [
    "pqs_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    pqs_dict[genome] = pqs_exec(na_seqs[genome])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_pqs_data(df):\n",
    "    df = df.sort_values([\"end\", \"start\"], ascending=True)\n",
    "\n",
    "    # Count length with overlapping positions\n",
    "    pos = 0\n",
    "    length = 0\n",
    "    for i, row in df.iterrows():\n",
    "        if row.start > pos:\n",
    "            length += row.end - row.start + 1\n",
    "            pos = row.end\n",
    "        elif row.start < pos and row.end > pos:\n",
    "            length += row.end - pos + 1\n",
    "            pos = row.end\n",
    "    \n",
    "    return [len(df), length]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_billingiae</th>\n",
       "      <th>E_gerundensis</th>\n",
       "      <th>E_persicina</th>\n",
       "      <th>E_pyrifoliae</th>\n",
       "      <th>E_tasmaniensis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Number of quadruplex regions (with overlaps)</th>\n",
       "      <td>445</td>\n",
       "      <td>212</td>\n",
       "      <td>913</td>\n",
       "      <td>445</td>\n",
       "      <td>508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Quadruplex length</th>\n",
       "      <td>12914</td>\n",
       "      <td>6138</td>\n",
       "      <td>27800</td>\n",
       "      <td>13259</td>\n",
       "      <td>15058</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              E_billingiae  E_gerundensis  \\\n",
       "Number of quadruplex regions (with overlaps)           445            212   \n",
       "Quadruplex length                                    12914           6138   \n",
       "\n",
       "                                              E_persicina  E_pyrifoliae  \\\n",
       "Number of quadruplex regions (with overlaps)          913           445   \n",
       "Quadruplex length                                   27800         13259   \n",
       "\n",
       "                                              E_tasmaniensis  \n",
       "Number of quadruplex regions (with overlaps)             508  \n",
       "Quadruplex length                                      15058  "
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pqs_stats = dict()\n",
    "for genome in genomes_list:\n",
    "    pqs_stats[genome] = process_pqs_data(pqs_dict[genome])\n",
    "\n",
    "pd.DataFrame(pqs_stats, index=[\"Number of quadruplex regions (with overlaps)\", \"Quadruplex length\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Рисуем гистограммы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "for genome in genomes_list:\n",
    "    sns.histplot(data=pqs_dict[genome], x=\"score\", bins=30, log_scale=False)\n",
    "    plt.title(genome.replace('_', '. ')+\" Quadruplex Scores\")\n",
    "    plt.savefig(f\"histograms/{genome}_quadruplex.png\", dpi=400)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Сохраняем данные и создаём .bed файлы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir -p pqs_result\n",
    "for genome in genomes_list:\n",
    "    pqs_dict[genome].to_csv(f\"pqs_result/{genome}.pqs.txt\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    df = pqs_dict[genome].copy()\n",
    "    bed_df = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": [genome] * len(df),\n",
    "            \"chromStart\": df.start,\n",
    "            \"chromEnd\": df.end,\n",
    "            \"name\": df.index,\n",
    "            \"score\": df.score,\n",
    "            \"strand\": df.strand\n",
    "        }\n",
    "    )\n",
    "\n",
    "    bed_df.to_csv(f\"pqs_result/{genome}.pqs.bed\", index=None, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Объединяем пересечения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools merge -c 5,6 -o max,distinct -i pqs_result/{genome}.pqs.bed > pqs_result/{genome}.pqs.merged.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ассоциируем предсказанные участки квадруплексов с промоторами генов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools intersect -a TSS/{genome}.slop.bed -b pqs_result/{genome}.pqs.merged.bed -wb > intersect/{genome}_quadruplex.inter.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_pqs_df(genome):\n",
    "    pqs = pd.read_csv(f\"intersect/{genome}_quadruplex.inter.bed\", sep='\\t', \n",
    "                names=[\"Sp1\", \"St1\", \"En1\", \"Protein\", \"Sp2\", \"Start\", \"End\", f\"Score_{genome}\", \"Strand\"]\n",
    "                )[[\"Protein\", f\"Score_{genome}\", \"Start\", \"End\", \"Strand\"]]\n",
    "    pqs.sort_values(f\"Score_{genome}\", ascending=False, inplace=True)\n",
    "    # Drop duplicate proteins, keep maximal score\n",
    "    pqs.drop_duplicates(subset='Protein', inplace=True)\n",
    "    pqs.set_index(\"Protein\", inplace=True)\n",
    "    return pqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "      <th>E_billingiae_protein.score</th>\n",
       "      <th>E_gerundensis_protein.score</th>\n",
       "      <th>E_persicina_protein.score</th>\n",
       "      <th>E_pyrifoliae_protein.score</th>\n",
       "      <th>E_tasmaniensis_protein.score</th>\n",
       "      <th>E_billingiae_protein.quadro</th>\n",
       "      <th>E_gerundensis_protein.quadro</th>\n",
       "      <th>E_persicina_protein.quadro</th>\n",
       "      <th>E_pyrifoliae_protein.quadro</th>\n",
       "      <th>E_tasmaniensis_protein.quadro</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.982</td>\n",
       "      <td>WP_013204225.1</td>\n",
       "      <td>WP_067427185.1</td>\n",
       "      <td>WP_222681877.1</td>\n",
       "      <td>WP_015899125.1</td>\n",
       "      <td>WP_012442922.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28780.500</td>\n",
       "      <td>766.6232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_173363027.1</td>\n",
       "      <td>WP_067429596.1</td>\n",
       "      <td>WP_174522410.1</td>\n",
       "      <td>WP_104945051.1</td>\n",
       "      <td>WP_012441756.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2943.461</td>\n",
       "      <td>2943.4610</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1790.574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013201311.1</td>\n",
       "      <td>WP_225700887.1</td>\n",
       "      <td>WP_062748436.1</td>\n",
       "      <td>WP_012668725.1</td>\n",
       "      <td>WP_012442032.1</td>\n",
       "      <td>908.3955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013200102.1</td>\n",
       "      <td>WP_067426621.1</td>\n",
       "      <td>WP_062746920.1</td>\n",
       "      <td>WP_012666468.1</td>\n",
       "      <td>WP_012439841.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1733.503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>783.823</td>\n",
       "      <td>865.283</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013203471.1</td>\n",
       "      <td>WP_067433095.1</td>\n",
       "      <td>WP_245006712.1</td>\n",
       "      <td>WP_012667256.1</td>\n",
       "      <td>WP_193345526.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          5      5       0.982           WP_013204225.1   \n",
       "1          5      5       0.981           WP_173363027.1   \n",
       "2          5      5       0.981           WP_013201311.1   \n",
       "3          5      5       0.981           WP_013200102.1   \n",
       "4          5      5       0.981           WP_013203471.1   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0            WP_067427185.1          WP_222681877.1           WP_015899125.1   \n",
       "1            WP_067429596.1          WP_174522410.1           WP_104945051.1   \n",
       "2            WP_225700887.1          WP_062748436.1           WP_012668725.1   \n",
       "3            WP_067426621.1          WP_062746920.1           WP_012666468.1   \n",
       "4            WP_067433095.1          WP_245006712.1           WP_012667256.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  E_billingiae_protein.score  \\\n",
       "0             WP_012442922.1                         NaN   \n",
       "1             WP_012441756.1                         NaN   \n",
       "2             WP_012442032.1                    908.3955   \n",
       "3             WP_012439841.1                         NaN   \n",
       "4             WP_193345526.1                         NaN   \n",
       "\n",
       "   E_gerundensis_protein.score  E_persicina_protein.score  \\\n",
       "0                    28780.500                   766.6232   \n",
       "1                     2943.461                  2943.4610   \n",
       "2                          NaN                        NaN   \n",
       "3                     1733.503                        NaN   \n",
       "4                          NaN                        NaN   \n",
       "\n",
       "   E_pyrifoliae_protein.score  E_tasmaniensis_protein.score  \\\n",
       "0                         NaN                           NaN   \n",
       "1                         NaN                      1790.574   \n",
       "2                         NaN                           NaN   \n",
       "3                     783.823                       865.283   \n",
       "4                         NaN                           NaN   \n",
       "\n",
       "   E_billingiae_protein.quadro  E_gerundensis_protein.quadro  \\\n",
       "0                          NaN                           NaN   \n",
       "1                          NaN                           NaN   \n",
       "2                          NaN                           NaN   \n",
       "3                          NaN                           NaN   \n",
       "4                          NaN                           NaN   \n",
       "\n",
       "   E_persicina_protein.quadro  E_pyrifoliae_protein.quadro  \\\n",
       "0                         NaN                          NaN   \n",
       "1                         NaN                          NaN   \n",
       "2                         NaN                          NaN   \n",
       "3                         NaN                          NaN   \n",
       "4                         NaN                          NaN   \n",
       "\n",
       "   E_tasmaniensis_protein.quadro  \n",
       "0                            NaN  \n",
       "1                            NaN  \n",
       "2                            NaN  \n",
       "3                            NaN  \n",
       "4                            NaN  "
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for genome in genomes_list:\n",
    "    pqs = read_pqs_df(genome)[f\"Score_{genome}\"]\n",
    "    list_of_scores = [float(pqs.loc[prot]) if prot in list(pqs.index) else None for prot in clusters[f\"{genome}_protein.faa\"]]\n",
    "    clusters[f\"{genome}_protein.quadro\"] = list_of_scores\n",
    "clusters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "      <th>E_billingiae_protein.score</th>\n",
       "      <th>E_gerundensis_protein.score</th>\n",
       "      <th>E_persicina_protein.score</th>\n",
       "      <th>E_pyrifoliae_protein.score</th>\n",
       "      <th>E_tasmaniensis_protein.score</th>\n",
       "      <th>E_billingiae_protein.quadro</th>\n",
       "      <th>E_gerundensis_protein.quadro</th>\n",
       "      <th>E_persicina_protein.quadro</th>\n",
       "      <th>E_pyrifoliae_protein.quadro</th>\n",
       "      <th>E_tasmaniensis_protein.quadro</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1143</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.929</td>\n",
       "      <td>WP_013200491.1</td>\n",
       "      <td>WP_225701318.1</td>\n",
       "      <td>WP_062745668.1</td>\n",
       "      <td>WP_012666867.1</td>\n",
       "      <td>WP_012442663.1</td>\n",
       "      <td>612.3848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>612.3848</td>\n",
       "      <td>3428.5290</td>\n",
       "      <td>3428.529</td>\n",
       "      <td>47.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>76.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1622</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.897</td>\n",
       "      <td>WP_013203132.1</td>\n",
       "      <td>WP_187485638.1</td>\n",
       "      <td>WP_062742990.1</td>\n",
       "      <td>WP_104945108.1</td>\n",
       "      <td>WP_012440908.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2188</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.298</td>\n",
       "      <td>WP_013203557.1</td>\n",
       "      <td>WP_067432667.1</td>\n",
       "      <td>WP_062744220.1</td>\n",
       "      <td>WP_012669075.1</td>\n",
       "      <td>WP_012442422.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13713.99</td>\n",
       "      <td>NaN</td>\n",
       "      <td>880.2792</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>91.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "1143          5      5       0.929           WP_013200491.1   \n",
       "1622          5      5       0.897           WP_013203132.1   \n",
       "2188          5      5       0.298           WP_013203557.1   \n",
       "\n",
       "     E_gerundensis_protein.faa E_persicina_protein.faa  \\\n",
       "1143            WP_225701318.1          WP_062745668.1   \n",
       "1622            WP_187485638.1          WP_062742990.1   \n",
       "2188            WP_067432667.1          WP_062744220.1   \n",
       "\n",
       "     E_pyrifoliae_protein.faa E_tasmaniensis_protein.faa  \\\n",
       "1143           WP_012666867.1             WP_012442663.1   \n",
       "1622           WP_104945108.1             WP_012440908.1   \n",
       "2188           WP_012669075.1             WP_012442422.1   \n",
       "\n",
       "      E_billingiae_protein.score  E_gerundensis_protein.score  \\\n",
       "1143                    612.3848                          NaN   \n",
       "1622                         NaN                          NaN   \n",
       "2188                         NaN                     13713.99   \n",
       "\n",
       "      E_persicina_protein.score  E_pyrifoliae_protein.score  \\\n",
       "1143                   612.3848                   3428.5290   \n",
       "1622                   883.5764                         NaN   \n",
       "2188                        NaN                    880.2792   \n",
       "\n",
       "      E_tasmaniensis_protein.score  E_billingiae_protein.quadro  \\\n",
       "1143                      3428.529                         47.0   \n",
       "1622                           NaN                         57.0   \n",
       "2188                           NaN                         59.0   \n",
       "\n",
       "      E_gerundensis_protein.quadro  E_persicina_protein.quadro  \\\n",
       "1143                          48.0                        55.0   \n",
       "1622                           NaN                        53.0   \n",
       "2188                           NaN                        53.0   \n",
       "\n",
       "      E_pyrifoliae_protein.quadro  E_tasmaniensis_protein.quadro  \n",
       "1143                         76.0                           76.0  \n",
       "1622                         49.0                           52.0  \n",
       "2188                         97.0                           91.0  "
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen = clusters.dropna(subset=clusters.columns[-5:], how=\"all\", thresh=4)\n",
    "chosen"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Визуализируем"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dna_features_viewer import GraphicFeature, GraphicRecord\n",
    "quadro_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    features_dict[genome].set_index(\"product_accession\", drop=False, inplace=True)\n",
    "    quadro_dict[genome] = read_pqs_df(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 1\n",
    "for k, row in chosen.iterrows():\n",
    "    rows_num = len(row.iloc[-5:].dropna())\n",
    "    fig, axes = plt.subplots(nrows=rows_num, ncols=1,figsize=(12, 6*rows_num/5))\n",
    "    fig.suptitle(f'Cluster {counter}')\n",
    "    axes_count = 0\n",
    "\n",
    "    for i, genome in enumerate(genomes_list):\n",
    "        quad = quadro_dict[genome]\n",
    "        features = features_dict[genome]\n",
    "\n",
    "        prot = row.loc[genome + '_protein.faa']\n",
    "        if not prot in quad.index:\n",
    "            continue\n",
    "\n",
    "        quad_label = \"Quadruplex, Score=\" + str(round(quad.loc[prot][f\"Score_{genome}\"]))\n",
    "        coords = [quad.loc[prot, \"Start\"], quad.loc[prot, \"End\"],\n",
    "                  features.loc[prot, \"start\"], features.loc[prot, \"end\"]]\n",
    "        minimum = min(coords)\n",
    "\n",
    "        strand_q = +1 if quad.loc[prot, \"Strand\"]==\"+\" else -1\n",
    "        strand_f = +1 if features.loc[prot, \"strand\"]==\"+\" else -1\n",
    "\n",
    "\n",
    "        features=[\n",
    "        GraphicFeature(start=coords[0]-minimum+26, end=coords[1]-minimum+26, \n",
    "                   strand=strand_q, color=\"#0057b8\", label=quad_label),\n",
    "        GraphicFeature(start=coords[2]-minimum+26, end=coords[3]-minimum+26, \n",
    "                   strand=strand_f, color=\"#ffd700\", label=genome + \": \" + prot),\n",
    "                   ]\n",
    "        record = GraphicRecord(sequence_length=max(coords)-minimum+50, features=features)\n",
    "        record.plot(ax=axes[axes_count])\n",
    "        axes_count += 1\n",
    "    plt.savefig(f'pictures/Cluster_{counter}_quadruplex.png', dpi=500)\n",
    "    counter += 1"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
