{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Обязательная часть"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Скачивание файлов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! sudo apt install clustalw bedtools proteinortho"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "base = \"http://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/\"\n",
    "paths = {\n",
    "    \"E_billingiae\": \"000/196/615/GCF_000196615.1_ASM19661v1\",\n",
    "    \"E_gerundensis\": \"020/342/335/GCF_020342335.1_ASM2034233v1\",\n",
    "    \"E_persicina\": \"019/844/095/GCF_019844095.1_ASM1984409v1\",\n",
    "    \"E_pyrifoliae\": \"002/952/315/GCF_002952315.1_ASM295231v1\",\n",
    "    \"E_tasmaniensis\": \"000/026/185/GCF_000026185.1_ASM2618v1\"\n",
    "}\n",
    "genomic_end = \"_genomic.fna.gz\"\n",
    "feature_end = \"_feature_table.txt.gz\"\n",
    "protein_end = \"_protein.faa.gz\"\n",
    "\n",
    "! mkdir -p genomes histograms zhunt_result TSS intersect ortho clusters/raw clusters/aligned pictures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in paths.keys():\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{genomic_end} -O 'genomes/{key}{genomic_end}'\")\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{feature_end} -O 'genomes/{key}{feature_end}'\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{genomic_end}\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{feature_end}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in paths.keys():\n",
    "    os.system(f\"wget -c -nv {base}{paths[key]}/{paths[key].split('/')[-1]}{protein_end} -O 'genomes/{key}{protein_end}'\")\n",
    "    os.system(f\"gzip -f -d genomes/{key}{protein_end}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Анализ аннотированных геномов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_list = list(paths.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annot_analyse(genome):\n",
    "    features = pd.read_csv(f\"genomes/{genome}_feature_table.txt\", sep='\\t')\n",
    "    features = features.loc[\n",
    "        (features['# feature']=='gene') & \n",
    "        (features['seq_type']=='chromosome')\n",
    "    ]\n",
    "    seqs = [seq_record for seq_record in SeqIO.parse(f\"genomes/{genome}_genomic.fna\", \"fasta\")]\n",
    "    # Choose complete genome, remove plasmids\n",
    "\n",
    "    genome_len = max(list(map(lambda x: len(x.seq), seqs)))\n",
    "    total_len = sum(list(map(lambda x: len(x.seq), seqs)))\n",
    "\n",
    "    total_genes_len = features[\"feature_interval_length\"].sum()\n",
    "\n",
    "    return [len(features), round(total_genes_len/genome_len*100, 2), genome_len, total_len, len(seqs)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_billingiae</th>\n",
       "      <th>E_gerundensis</th>\n",
       "      <th>E_persicina</th>\n",
       "      <th>E_pyrifoliae</th>\n",
       "      <th>E_tasmaniensis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Genes number</th>\n",
       "      <td>4703.00</td>\n",
       "      <td>3458.00</td>\n",
       "      <td>4311.00</td>\n",
       "      <td>3746.00</td>\n",
       "      <td>3569.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome coverage</th>\n",
       "      <td>88.99</td>\n",
       "      <td>88.19</td>\n",
       "      <td>88.61</td>\n",
       "      <td>86.27</td>\n",
       "      <td>86.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome length</th>\n",
       "      <td>5100167.00</td>\n",
       "      <td>3748909.00</td>\n",
       "      <td>4664605.00</td>\n",
       "      <td>4027225.00</td>\n",
       "      <td>3883467.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genome + plasmid length</th>\n",
       "      <td>5372268.00</td>\n",
       "      <td>4437416.00</td>\n",
       "      <td>4802925.00</td>\n",
       "      <td>4075681.00</td>\n",
       "      <td>4067864.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Number of sequences</th>\n",
       "      <td>3.00</td>\n",
       "      <td>3.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>2.00</td>\n",
       "      <td>6.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         E_billingiae  E_gerundensis  E_persicina  \\\n",
       "Genes number                  4703.00        3458.00      4311.00   \n",
       "Genome coverage                 88.99          88.19        88.61   \n",
       "Genome length              5100167.00     3748909.00   4664605.00   \n",
       "Genome + plasmid length    5372268.00     4437416.00   4802925.00   \n",
       "Number of sequences              3.00           3.00         2.00   \n",
       "\n",
       "                         E_pyrifoliae  E_tasmaniensis  \n",
       "Genes number                  3746.00         3569.00  \n",
       "Genome coverage                 86.27           86.71  \n",
       "Genome length              4027225.00      3883467.00  \n",
       "Genome + plasmid length    4075681.00      4067864.00  \n",
       "Number of sequences              2.00            6.00  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    annot_dict[genome] = annot_analyse(genome)\n",
    "\n",
    "pd.DataFrame(annot_dict, \n",
    "index=[\"Genes number\", \"Genome coverage\", \n",
    "\"Genome length\", \"Genome + plasmid length\", \"Number of sequences\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Предсказание участков Z-DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system(\"wget -c -nv https://raw.githubusercontent.com/vanya-antonov/hse22-project/main/zhunt3-alan.c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system(\"gcc zhunt3-alan.c -lm -o zhunt3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from subprocess import DEVNULL, PIPE\n",
    "\n",
    "ZH_EXECUTABLE = Path(\"../hse22_project/zhunt3\")\n",
    "assert ZH_EXECUTABLE.is_file()\n",
    "\n",
    "\n",
    "def zhunt(query: str, windowsize: int = 6, minsize: int = 3, maxsize: int = 6):\n",
    "    assert set(query).issubset({\"A\", \"C\", \"G\", \"T\", \"N\"})\n",
    "    fd, temp = tempfile.mkstemp()\n",
    "    os.close(fd)\n",
    "    with open(temp, 'w') as stream:\n",
    "        stream.write(query)\n",
    "\n",
    "    subprocess.run(\n",
    "        [ZH_EXECUTABLE, \n",
    "         str(windowsize), str(minsize), str(maxsize), temp],\n",
    "        check=True, stdout=PIPE, stderr=DEVNULL,\n",
    "        input=query, encoding='ascii'\n",
    "    )\n",
    "    with open(temp + \".Z-SCORE\", 'r') as stream:\n",
    "        df = pd.read_csv(stream,\n",
    "                         names=['Start', 'End', 'nu-1', 'nu-2', 'nu-3', \n",
    "                                'ZH-Score', 'Sequence', 'Conformation'],\n",
    "                         skiprows=1, sep='\\s+')\n",
    "    os.remove(temp)\n",
    "    os.remove(temp + \".Z-SCORE\")\n",
    "    df = df.loc[df[\"ZH-Score\"] > 500]\n",
    "    return df[['Start', 'End', 'ZH-Score', 'Sequence', 'Conformation']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "def choose_max_seq(genome):\n",
    "    seqs = [str(seq_record.seq) for seq_record in SeqIO.parse(f\"genomes/{genome}_genomic.fna\", \"fasta\")]\n",
    "    # Choose complete genome, remove plasmids\n",
    "    max_scaffold = max(seqs, key=len)\n",
    "    return max_scaffold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "zhunt_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    zhunt_dict[genome] = zhunt(choose_max_seq(genome))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ZH-Score</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Conformation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>734</td>\n",
       "      <td>742</td>\n",
       "      <td>766.6232</td>\n",
       "      <td>cgtgcgtg</td>\n",
       "      <td>ASASASAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1677</th>\n",
       "      <td>1678</td>\n",
       "      <td>1690</td>\n",
       "      <td>1356.2120</td>\n",
       "      <td>gggcgtgcccgc</td>\n",
       "      <td>SASASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1679</th>\n",
       "      <td>1680</td>\n",
       "      <td>1690</td>\n",
       "      <td>1201.6710</td>\n",
       "      <td>gcgtgcccgc</td>\n",
       "      <td>SASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1744</th>\n",
       "      <td>1745</td>\n",
       "      <td>1753</td>\n",
       "      <td>516.0707</td>\n",
       "      <td>ttgcgcgc</td>\n",
       "      <td>SASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1746</th>\n",
       "      <td>1747</td>\n",
       "      <td>1753</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>gcgcgc</td>\n",
       "      <td>SASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5096951</th>\n",
       "      <td>5096952</td>\n",
       "      <td>5096962</td>\n",
       "      <td>2183.5740</td>\n",
       "      <td>cgtgcgtgtg</td>\n",
       "      <td>ASASASASAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5097858</th>\n",
       "      <td>5097859</td>\n",
       "      <td>5097871</td>\n",
       "      <td>717.9641</td>\n",
       "      <td>aagcgtatgcgc</td>\n",
       "      <td>SASASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5097860</th>\n",
       "      <td>5097861</td>\n",
       "      <td>5097871</td>\n",
       "      <td>948.8341</td>\n",
       "      <td>gcgtatgcgc</td>\n",
       "      <td>SASASASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5098505</th>\n",
       "      <td>5098506</td>\n",
       "      <td>5098512</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>gcgcgc</td>\n",
       "      <td>SASASA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5099288</th>\n",
       "      <td>5099289</td>\n",
       "      <td>5099297</td>\n",
       "      <td>908.3955</td>\n",
       "      <td>gtgcgcgt</td>\n",
       "      <td>SASASASA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29749 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Start      End   ZH-Score      Sequence  Conformation\n",
       "733          734      742   766.6232      cgtgcgtg      ASASASAS\n",
       "1677        1678     1690  1356.2120  gggcgtgcccgc  SASASASASASA\n",
       "1679        1680     1690  1201.6710    gcgtgcccgc    SASASASASA\n",
       "1744        1745     1753   516.0707      ttgcgcgc      SASASASA\n",
       "1746        1747     1753   883.5764        gcgcgc        SASASA\n",
       "...          ...      ...        ...           ...           ...\n",
       "5096951  5096952  5096962  2183.5740    cgtgcgtgtg    ASASASASAS\n",
       "5097858  5097859  5097871   717.9641  aagcgtatgcgc  SASASASASASA\n",
       "5097860  5097861  5097871   948.8341    gcgtatgcgc    SASASASASA\n",
       "5098505  5098506  5098512   883.5764        gcgcgc        SASASA\n",
       "5099288  5099289  5099297   908.3955      gtgcgcgt      SASASASA\n",
       "\n",
       "[29749 rows x 5 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhunt_dict[genomes_list[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_zhunt_data(df):\n",
    "    df = df.sort_values([\"End\", \"Start\"], ascending=True)\n",
    "\n",
    "    # Count length with overlapping positions\n",
    "    pos = 0\n",
    "    length = 0\n",
    "    for i, row in df.iterrows():\n",
    "        if row.Start > pos:\n",
    "            length += row.End - row.Start + 1\n",
    "            pos = row.End\n",
    "        elif row.Start < pos and row.End > pos:\n",
    "            length += row.End - pos + 1\n",
    "            pos = row.End\n",
    "    \n",
    "    return [len(df), length]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>E_billingiae</th>\n",
       "      <th>E_gerundensis</th>\n",
       "      <th>E_persicina</th>\n",
       "      <th>E_pyrifoliae</th>\n",
       "      <th>E_tasmaniensis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Number of Z-DNA regions (with overlaps)</th>\n",
       "      <td>29749</td>\n",
       "      <td>33366</td>\n",
       "      <td>25907</td>\n",
       "      <td>25346</td>\n",
       "      <td>26301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z-DNA length</th>\n",
       "      <td>119602</td>\n",
       "      <td>131756</td>\n",
       "      <td>107026</td>\n",
       "      <td>102084</td>\n",
       "      <td>103424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         E_billingiae  E_gerundensis  \\\n",
       "Number of Z-DNA regions (with overlaps)         29749          33366   \n",
       "Z-DNA length                                   119602         131756   \n",
       "\n",
       "                                         E_persicina  E_pyrifoliae  \\\n",
       "Number of Z-DNA regions (with overlaps)        25907         25346   \n",
       "Z-DNA length                                  107026        102084   \n",
       "\n",
       "                                         E_tasmaniensis  \n",
       "Number of Z-DNA regions (with overlaps)           26301  \n",
       "Z-DNA length                                     103424  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zhunt_stats = dict()\n",
    "for genome in genomes_list:\n",
    "    zhunt_stats[genome] = process_zhunt_data(zhunt_dict[genome])\n",
    "\n",
    "pd.DataFrame(zhunt_stats, index=[\"Number of Z-DNA regions (with overlaps)\", \"Z-DNA length\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Рисуем гистограммы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "for genome in genomes_list:\n",
    "    sns.histplot(data=zhunt_dict[genome], x=\"ZH-Score\", bins=30, log_scale=True)\n",
    "    plt.title(genome.replace('_', '. ')+\" ZH-Scores\")\n",
    "    plt.savefig(f\"histograms/{genome}.png\", dpi=400)\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Сохраняем данные и создаём .bed файлы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    zhunt_dict[genome].to_csv(f\"zhunt_result/{genome}.zhunt.txt\", index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    df = zhunt_dict[genomes_list[0]].copy()\n",
    "    bed_df = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": [genome] * len(df),\n",
    "            \"chromStart\": df.Start,\n",
    "            \"chromEnd\": df.End,\n",
    "            \"name\": df.index,\n",
    "            \"score\": df[\"ZH-Score\"]\n",
    "        }\n",
    "    )\n",
    "\n",
    "    bed_df.to_csv(f\"zhunt_result/{genome}.zhunt.bed\", index=None, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Объединяем пересечения"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools merge -c 5 -o max -i zhunt_result/{genome}.zhunt.bed > zhunt_result/{genome}.zhunt.merged.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ассоциируем предсказанные участки Z-DNA с промоторами генов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "features_dict = dict()\n",
    "\n",
    "for genome in genomes_list:\n",
    "    features = pd.read_csv(f\"genomes/{genome}_feature_table.txt\", sep='\\t')\n",
    "    features = features.loc[\n",
    "        (features['seq_type']=='chromosome')\n",
    "    ]\n",
    "    locuses = list(np.unique(features.locus_tag))\n",
    "    df = features.copy()\n",
    "\n",
    "    counter = 1\n",
    "    for locus in locuses:\n",
    "        prod = list(features['product_accession'].loc[features.locus_tag == locus].dropna())\n",
    "        if prod == []:\n",
    "            prod = [\"RNA\" + str(counter)]\n",
    "            counter += 1\n",
    "            df['product_accession'].loc[df.locus_tag == locus] = prod[0]\n",
    "        elif len(prod) == 1:\n",
    "            df['product_accession'].loc[df.locus_tag == locus] = prod[0]\n",
    "        else:\n",
    "            print(prod)\n",
    "    df = df.loc[df['# feature'] == 'gene']\n",
    "    df = df[[\"# feature\", \"class\" , \"start\", \"end\", \"strand\", \"product_accession\", \"locus_tag\"]]\n",
    "    features_dict[genome] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    features_dict[genome]['TSS'] = features_dict[genome].apply(lambda x: x.start if x.strand == \"+\" else x.end, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    df = features_dict[genome].copy()\n",
    "    bed_df = pd.DataFrame(\n",
    "        {\n",
    "            \"chrom\": [genome] * len(df),\n",
    "            \"chromStart\": df.TSS,\n",
    "            \"chromEnd\": df.TSS,\n",
    "            \"name\": df.product_accession\n",
    "        }\n",
    "    )\n",
    "\n",
    "    bed_df.to_csv(f\"TSS/{genome}.bed\", index=None, header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(\n",
    "    {\n",
    "        \"Species\": genomes_list,\n",
    "        \"Len\": [annot_dict[genome][2] for genome in genomes_list]\n",
    "    }\n",
    ").to_csv(f\"TSS/my.genomes\", index=None, header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "for genome in genomes_list:\n",
    "    os.system(f\"bedtools slop -g TSS/my.genomes -i TSS/{genome}.bed -b 100 > TSS/{genome}.slop.bed\")\n",
    "    os.system(f\"bedtools intersect -a TSS/{genome}.slop.bed -b zhunt_result/{genome}.zhunt.merged.bed -wb > intersect/{genome}.inter.bed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Создаём кластеры"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "*****************************************************************\n",
      "\u001b[1;32mProteinortho\u001b[0m with PoFF version 6.0.14 - An orthology detection tool\n",
      "*****************************************************************\n",
      "Detected 8 available CPU threads (adjust this with -cpus), Detected 'diamond' version 0.9.30\n",
      "Checking input files.\n",
      "Checking genomes/E_billingiae_protein.faa... ok\n",
      "Checking genomes/E_gerundensis_protein.faa... ok\n",
      "Checking genomes/E_persicina_protein.faa... ok\n",
      "Checking genomes/E_pyrifoliae_protein.faa... ok\n",
      "Checking genomes/E_tasmaniensis_protein.faa... ok\n",
      "\n",
      "\u001b[1;32m**Step 1**\u001b[0m\n",
      "Generating indices.\n",
      "The database for 'genomes/E_billingiae_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_persicina_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_gerundensis_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_tasmaniensis_protein.faa' is present and will be used\n",
      "The database for 'genomes/E_pyrifoliae_protein.faa' is present and will be used\n",
      "\n",
      "\u001b[1;32m**Step 2**\u001b[0m using diamond \n",
      "Running blast analysis: 100% (10/10)                                           \n",
      "[OUTPUT] -> written to erwinia.blast-graph\n",
      "\n",
      "\u001b[1;32m**Step 3**\u001b[0m\n",
      "Clustering by similarity (Proteinortho mode) using up to 5883 MB of memory (75% of total memory) and 8 cpu core(s). Adjust this behaviour with the -mem option.\n",
      "Reading erwinia.blast-graph\n",
      "5 species\n",
      "16925 paired proteins\n",
      "27875 bidirectional edges\n",
      "Done                       \n",
      "[OUTPUT] -> Orthologous groups are written to erwinia.proteinortho.tsv\n",
      "You can extract the fasta files of each orthology group with 'proteinortho_grab_proteins.pl -tofiles erwinia.proteinortho.tsv genomes/E_billingiae_protein.faa genomes/E_persicina_protein.faa genomes/E_gerundensis_protein.faa genomes/E_tasmaniensis_protein.faa genomes/E_pyrifoliae_protein.faa'\n",
      " (Careful: This will generate a file foreach line in the file erwinia.proteinortho.tsv).\n",
      "[OUTPUT] -> Orthologous pairs are written to erwinia.proteinortho-graph\n",
      "[OUTPUT] -> Summary is written to erwinia.proteinortho-graph.summary\n",
      "[OUTPUT] -> Orthologous groups are written to erwinia.proteinortho.html\n",
      "\n",
      "\u001b[1;32mAll finished.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!proteinortho -project=erwinia genomes/*faa\n",
    "!mv erwinia* ortho/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Статистика по кластерам"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WP_013203652.1</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012440702.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669311.1</td>\n",
       "      <td>WP_012440279.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669308.1</td>\n",
       "      <td>WP_012440282.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669307.1</td>\n",
       "      <td>WP_012440283.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>*</td>\n",
       "      <td>WP_012669404.1</td>\n",
       "      <td>WP_012440317.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          2      2         1.0           WP_013203652.1   \n",
       "1          2      2         1.0                        *   \n",
       "2          2      2         1.0                        *   \n",
       "3          2      2         1.0                        *   \n",
       "4          2      2         1.0                        *   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0                         *                       *                        *   \n",
       "1                         *                       *           WP_012669311.1   \n",
       "2                         *                       *           WP_012669308.1   \n",
       "3                         *                       *           WP_012669307.1   \n",
       "4                         *                       *           WP_012669404.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  \n",
       "0             WP_012440702.1  \n",
       "1             WP_012440279.1  \n",
       "2             WP_012440282.1  \n",
       "3             WP_012440283.1  \n",
       "4             WP_012440317.1  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters = pd.read_csv(\"ortho/erwinia.proteinortho.tsv\", sep='\\t')\n",
    "clusters.sort_values(\"Alg.-Conn.\", ascending=False, inplace=True)\n",
    "clusters.reset_index(drop=True, inplace=True)\n",
    "clusters.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4337"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_of_clusters = len(clusters)\n",
    "number_of_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_theme()\n",
    "sns.countplot(data=clusters, x=\"# Species\")\n",
    "plt.title(\"Distribution of species in clusters\")\n",
    "#plt.xticks(ticks=[2, 3, 4, 5], labels=['2', '3', '4', '5'])\n",
    "plt.tight_layout()\n",
    "plt.savefig(f\"histograms/clusters.png\", dpi=400)\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Пересекаем кластеры и Z-DNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.982</td>\n",
       "      <td>WP_013204225.1</td>\n",
       "      <td>WP_067427185.1</td>\n",
       "      <td>WP_222681877.1</td>\n",
       "      <td>WP_015899125.1</td>\n",
       "      <td>WP_012442922.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_173363027.1</td>\n",
       "      <td>WP_067429596.1</td>\n",
       "      <td>WP_174522410.1</td>\n",
       "      <td>WP_104945051.1</td>\n",
       "      <td>WP_012441756.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013201311.1</td>\n",
       "      <td>WP_225700887.1</td>\n",
       "      <td>WP_062748436.1</td>\n",
       "      <td>WP_012668725.1</td>\n",
       "      <td>WP_012442032.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013200102.1</td>\n",
       "      <td>WP_067426621.1</td>\n",
       "      <td>WP_062746920.1</td>\n",
       "      <td>WP_012666468.1</td>\n",
       "      <td>WP_012439841.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013203471.1</td>\n",
       "      <td>WP_067433095.1</td>\n",
       "      <td>WP_245006712.1</td>\n",
       "      <td>WP_012667256.1</td>\n",
       "      <td>WP_193345526.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          5      5       0.982           WP_013204225.1   \n",
       "1          5      5       0.981           WP_173363027.1   \n",
       "2          5      5       0.981           WP_013201311.1   \n",
       "3          5      5       0.981           WP_013200102.1   \n",
       "4          5      5       0.981           WP_013203471.1   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0            WP_067427185.1          WP_222681877.1           WP_015899125.1   \n",
       "1            WP_067429596.1          WP_174522410.1           WP_104945051.1   \n",
       "2            WP_225700887.1          WP_062748436.1           WP_012668725.1   \n",
       "3            WP_067426621.1          WP_062746920.1           WP_012666468.1   \n",
       "4            WP_067433095.1          WP_245006712.1           WP_012667256.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  \n",
       "0             WP_012442922.1  \n",
       "1             WP_012441756.1  \n",
       "2             WP_012442032.1  \n",
       "3             WP_012439841.1  \n",
       "4             WP_193345526.1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters = pd.read_csv(\"ortho/erwinia.proteinortho.tsv\", sep='\\t')\n",
    "clusters = clusters.loc[clusters['# Species']==5]\n",
    "clusters.sort_values(\"Alg.-Conn.\", ascending=False, inplace=True)\n",
    "clusters.reset_index(drop=True, inplace=True)\n",
    "clusters.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_zdna_df(genome):\n",
    "    zdna = pd.read_csv(f\"intersect/{genome}.inter.bed\", sep='\\t', \n",
    "                names=[\"Sp1\", \"St1\", \"En1\", \"Protein\", \"Sp2\", \"Start\", \"End\", f\"Score_{genome}\"]\n",
    "                )[[\"Protein\", f\"Score_{genome}\", \"Start\", \"End\"]]\n",
    "    zdna.sort_values(f\"Score_{genome}\", ascending=False, inplace=True)\n",
    "    # Drop duplicate proteins, keep maximal score\n",
    "    zdna.drop_duplicates(subset='Protein', inplace=True)\n",
    "    zdna.set_index(\"Protein\", inplace=True)\n",
    "    return zdna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "      <th>E_billingiae_protein.score</th>\n",
       "      <th>E_gerundensis_protein.score</th>\n",
       "      <th>E_persicina_protein.score</th>\n",
       "      <th>E_pyrifoliae_protein.score</th>\n",
       "      <th>E_tasmaniensis_protein.score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.982</td>\n",
       "      <td>WP_013204225.1</td>\n",
       "      <td>WP_067427185.1</td>\n",
       "      <td>WP_222681877.1</td>\n",
       "      <td>WP_015899125.1</td>\n",
       "      <td>WP_012442922.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>860.1321</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_173363027.1</td>\n",
       "      <td>WP_067429596.1</td>\n",
       "      <td>WP_174522410.1</td>\n",
       "      <td>WP_104945051.1</td>\n",
       "      <td>WP_012441756.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>731.2843</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>3428.529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013201311.1</td>\n",
       "      <td>WP_225700887.1</td>\n",
       "      <td>WP_062748436.1</td>\n",
       "      <td>WP_012668725.1</td>\n",
       "      <td>WP_012442032.1</td>\n",
       "      <td>908.3955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3428.5290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013200102.1</td>\n",
       "      <td>WP_067426621.1</td>\n",
       "      <td>WP_062746920.1</td>\n",
       "      <td>WP_012666468.1</td>\n",
       "      <td>WP_012439841.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.981</td>\n",
       "      <td>WP_013203471.1</td>\n",
       "      <td>WP_067433095.1</td>\n",
       "      <td>WP_245006712.1</td>\n",
       "      <td>WP_012667256.1</td>\n",
       "      <td>WP_193345526.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>650.9198</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "0          5      5       0.982           WP_013204225.1   \n",
       "1          5      5       0.981           WP_173363027.1   \n",
       "2          5      5       0.981           WP_013201311.1   \n",
       "3          5      5       0.981           WP_013200102.1   \n",
       "4          5      5       0.981           WP_013203471.1   \n",
       "\n",
       "  E_gerundensis_protein.faa E_persicina_protein.faa E_pyrifoliae_protein.faa  \\\n",
       "0            WP_067427185.1          WP_222681877.1           WP_015899125.1   \n",
       "1            WP_067429596.1          WP_174522410.1           WP_104945051.1   \n",
       "2            WP_225700887.1          WP_062748436.1           WP_012668725.1   \n",
       "3            WP_067426621.1          WP_062746920.1           WP_012666468.1   \n",
       "4            WP_067433095.1          WP_245006712.1           WP_012667256.1   \n",
       "\n",
       "  E_tasmaniensis_protein.faa  E_billingiae_protein.score  \\\n",
       "0             WP_012442922.1                         NaN   \n",
       "1             WP_012441756.1                         NaN   \n",
       "2             WP_012442032.1                    908.3955   \n",
       "3             WP_012439841.1                         NaN   \n",
       "4             WP_193345526.1                         NaN   \n",
       "\n",
       "   E_gerundensis_protein.score  E_persicina_protein.score  \\\n",
       "0                          NaN                   860.1321   \n",
       "1                          NaN                   731.2843   \n",
       "2                          NaN                  3428.5290   \n",
       "3                          NaN                        NaN   \n",
       "4                          NaN                        NaN   \n",
       "\n",
       "   E_pyrifoliae_protein.score  E_tasmaniensis_protein.score  \n",
       "0                         NaN                           NaN  \n",
       "1                    883.5764                      3428.529  \n",
       "2                         NaN                           NaN  \n",
       "3                         NaN                           NaN  \n",
       "4                    650.9198                           NaN  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for genome in genomes_list:\n",
    "    zdna = read_zdna_df(genome)[f\"Score_{genome}\"]\n",
    "    list_of_zhscores = [float(zdna.loc[prot]) if prot in list(zdna.index) else None for prot in clusters[f\"{genome}_protein.faa\"]]\n",
    "    clusters[f\"{genome}_protein.score\"] = list_of_zhscores\n",
    "clusters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># Species</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Alg.-Conn.</th>\n",
       "      <th>E_billingiae_protein.faa</th>\n",
       "      <th>E_gerundensis_protein.faa</th>\n",
       "      <th>E_persicina_protein.faa</th>\n",
       "      <th>E_pyrifoliae_protein.faa</th>\n",
       "      <th>E_tasmaniensis_protein.faa</th>\n",
       "      <th>E_billingiae_protein.score</th>\n",
       "      <th>E_gerundensis_protein.score</th>\n",
       "      <th>E_persicina_protein.score</th>\n",
       "      <th>E_pyrifoliae_protein.score</th>\n",
       "      <th>E_tasmaniensis_protein.score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.966</td>\n",
       "      <td>WP_013200370.1</td>\n",
       "      <td>WP_225701291.1</td>\n",
       "      <td>WP_062748290.1</td>\n",
       "      <td>WP_014539587.1</td>\n",
       "      <td>WP_012442783.1</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>1915.4320</td>\n",
       "      <td>1264.1890</td>\n",
       "      <td>980.8116</td>\n",
       "      <td>2659.7900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.965</td>\n",
       "      <td>WP_013203580.1</td>\n",
       "      <td>WP_067433293.1</td>\n",
       "      <td>WP_062744233.1</td>\n",
       "      <td>WP_012669090.1</td>\n",
       "      <td>WP_012442437.1</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>818.2755</td>\n",
       "      <td>3428.5290</td>\n",
       "      <td>907.7149</td>\n",
       "      <td>663.3761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>930</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.936</td>\n",
       "      <td>WP_013200783.1</td>\n",
       "      <td>WP_225700991.1</td>\n",
       "      <td>WP_062747223.1</td>\n",
       "      <td>WP_012667145.1</td>\n",
       "      <td>WP_012440562.1</td>\n",
       "      <td>3428.5290</td>\n",
       "      <td>826.8355</td>\n",
       "      <td>8697.4600</td>\n",
       "      <td>1323.0540</td>\n",
       "      <td>783.8230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.934</td>\n",
       "      <td>WP_013201679.1</td>\n",
       "      <td>WP_067429626.1</td>\n",
       "      <td>WP_062743729.1</td>\n",
       "      <td>WP_012668438.1</td>\n",
       "      <td>WP_012441745.1</td>\n",
       "      <td>736.1451</td>\n",
       "      <td>3135.7890</td>\n",
       "      <td>13713.9900</td>\n",
       "      <td>13713.9900</td>\n",
       "      <td>883.5764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1067</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.932</td>\n",
       "      <td>WP_013203485.1</td>\n",
       "      <td>WP_003849035.1</td>\n",
       "      <td>WP_062744160.1</td>\n",
       "      <td>WP_004155938.1</td>\n",
       "      <td>WP_012442362.1</td>\n",
       "      <td>1737.6120</td>\n",
       "      <td>783.8230</td>\n",
       "      <td>980.8116</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>13713.9900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1544</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.905</td>\n",
       "      <td>WP_013202442.1</td>\n",
       "      <td>WP_225701617.1</td>\n",
       "      <td>WP_062746141.1</td>\n",
       "      <td>WP_014538862.1</td>\n",
       "      <td>WP_012441295.1</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>980.8116</td>\n",
       "      <td>6970.7950</td>\n",
       "      <td>2752.4470</td>\n",
       "      <td>883.5764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1845</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.855</td>\n",
       "      <td>WP_013202465.1</td>\n",
       "      <td>WP_225701614.1</td>\n",
       "      <td>WP_222682263.1</td>\n",
       "      <td>WP_012667919.1</td>\n",
       "      <td>WP_012441270.1</td>\n",
       "      <td>883.5764</td>\n",
       "      <td>915.9191</td>\n",
       "      <td>783.8230</td>\n",
       "      <td>2752.4470</td>\n",
       "      <td>883.5764</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1879</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0.843</td>\n",
       "      <td>WP_013200831.1</td>\n",
       "      <td>WP_209500512.1</td>\n",
       "      <td>WP_191927791.1</td>\n",
       "      <td>WP_012667211.1</td>\n",
       "      <td>WP_012440626.1</td>\n",
       "      <td>766.6232</td>\n",
       "      <td>38833.5800</td>\n",
       "      <td>766.6232</td>\n",
       "      <td>904.3200</td>\n",
       "      <td>8782.1570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      # Species  Genes  Alg.-Conn. E_billingiae_protein.faa  \\\n",
       "123           5      5       0.966           WP_013200370.1   \n",
       "132           5      5       0.965           WP_013203580.1   \n",
       "930           5      5       0.936           WP_013200783.1   \n",
       "990           5      5       0.934           WP_013201679.1   \n",
       "1067          5      5       0.932           WP_013203485.1   \n",
       "1544          5      5       0.905           WP_013202442.1   \n",
       "1845          5      5       0.855           WP_013202465.1   \n",
       "1879          5      5       0.843           WP_013200831.1   \n",
       "\n",
       "     E_gerundensis_protein.faa E_persicina_protein.faa  \\\n",
       "123             WP_225701291.1          WP_062748290.1   \n",
       "132             WP_067433293.1          WP_062744233.1   \n",
       "930             WP_225700991.1          WP_062747223.1   \n",
       "990             WP_067429626.1          WP_062743729.1   \n",
       "1067            WP_003849035.1          WP_062744160.1   \n",
       "1544            WP_225701617.1          WP_062746141.1   \n",
       "1845            WP_225701614.1          WP_222682263.1   \n",
       "1879            WP_209500512.1          WP_191927791.1   \n",
       "\n",
       "     E_pyrifoliae_protein.faa E_tasmaniensis_protein.faa  \\\n",
       "123            WP_014539587.1             WP_012442783.1   \n",
       "132            WP_012669090.1             WP_012442437.1   \n",
       "930            WP_012667145.1             WP_012440562.1   \n",
       "990            WP_012668438.1             WP_012441745.1   \n",
       "1067           WP_004155938.1             WP_012442362.1   \n",
       "1544           WP_014538862.1             WP_012441295.1   \n",
       "1845           WP_012667919.1             WP_012441270.1   \n",
       "1879           WP_012667211.1             WP_012440626.1   \n",
       "\n",
       "      E_billingiae_protein.score  E_gerundensis_protein.score  \\\n",
       "123                     883.5764                    1915.4320   \n",
       "132                     883.5764                     818.2755   \n",
       "930                    3428.5290                     826.8355   \n",
       "990                     736.1451                    3135.7890   \n",
       "1067                   1737.6120                     783.8230   \n",
       "1544                    883.5764                     980.8116   \n",
       "1845                    883.5764                     915.9191   \n",
       "1879                    766.6232                   38833.5800   \n",
       "\n",
       "      E_persicina_protein.score  E_pyrifoliae_protein.score  \\\n",
       "123                   1264.1890                    980.8116   \n",
       "132                   3428.5290                    907.7149   \n",
       "930                   8697.4600                   1323.0540   \n",
       "990                  13713.9900                  13713.9900   \n",
       "1067                   980.8116                    883.5764   \n",
       "1544                  6970.7950                   2752.4470   \n",
       "1845                   783.8230                   2752.4470   \n",
       "1879                   766.6232                    904.3200   \n",
       "\n",
       "      E_tasmaniensis_protein.score  \n",
       "123                      2659.7900  \n",
       "132                       663.3761  \n",
       "930                       783.8230  \n",
       "990                       883.5764  \n",
       "1067                    13713.9900  \n",
       "1544                      883.5764  \n",
       "1845                      883.5764  \n",
       "1879                     8782.1570  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chosen = clusters.dropna(subset=clusters.columns[-5:])\n",
    "chosen"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Выравниваем белки"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_seqs = dict()\n",
    "for genome in genomes_list:\n",
    "    aa_seqs[genome] = SeqIO.to_dict(SeqIO.parse(f\"genomes/{genome}_protein.faa\", \"fasta\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 1\n",
    "\n",
    "for i, row in chosen.iterrows():\n",
    "    seqs_in_cluster = []\n",
    "    for genome in genomes_list:\n",
    "        protein_id = row[f\"{genome}_protein.faa\"]\n",
    "        seqs_in_cluster.append(aa_seqs[genome][protein_id])\n",
    "    SeqIO.write(seqs_in_cluster, f\"clusters/raw/cluster_{counter}.faa\", \"fasta\")\n",
    "    counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cl in range(1,9):\n",
    "    os.system(f\"clustalw -align -type=PROTEIN -infile='clusters/raw/cluster_{cl}.faa' -OUTFILE='clusters/aligned/cluster_{cl}.aln'\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Визуализируем"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dna_features_viewer import GraphicFeature, GraphicRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "zdna_dict = dict()\n",
    "for genome in genomes_list:\n",
    "    features_dict[genome].set_index(\"product_accession\", drop=False, inplace=True)\n",
    "    zdna_dict[genome] = read_zdna_df(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counter = 1\n",
    "for k, row in chosen.iterrows():\n",
    "    fig, axes = plt.subplots(nrows=5, ncols=1,figsize=(12, 6))\n",
    "    fig.suptitle(f'Cluster {counter}')\n",
    "\n",
    "    for i, genome in enumerate(genomes_list):\n",
    "        zdna = zdna_dict[genome]\n",
    "        features = features_dict[genome]\n",
    "\n",
    "        prot = row.loc[genome + '_protein.faa']\n",
    "        Z_DNA_label = \"Z-DNA, ZH-Score=\" + str(round(zdna.loc[prot][f\"Score_{genome}\"]))\n",
    "        coords = [zdna.loc[prot, \"Start\"], zdna.loc[prot, \"End\"],\n",
    "                  features.loc[prot, \"start\"], features.loc[prot, \"end\"]]\n",
    "        minimum = min(coords)\n",
    "\n",
    "        strand = +1 if features.loc[prot, \"strand\"]==\"+\" else -1\n",
    "\n",
    "        features=[\n",
    "        GraphicFeature(start=coords[0]-minimum+26, end=coords[1]-minimum+26, \n",
    "                   strand=+1, color=\"#0057b8\", label=Z_DNA_label),\n",
    "        GraphicFeature(start=coords[2]-minimum+26, end=coords[3]-minimum+26, \n",
    "                   strand=strand, color=\"#ffd700\", label=prot),\n",
    "                   ]\n",
    "        record = GraphicRecord(sequence_length=max(coords)-minimum+50, features=features)\n",
    "        record.plot(ax=axes[i])\n",
    "    plt.savefig(f'pictures/Cluster_{counter}.png', dpi=500)\n",
    "    counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Бонусная часть"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Поиск G-квадруплексов"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "na_seqs = dict()\n",
    "for genome in genomes_list:\n",
    "    na_seqs[genome] = choose_max_seq(genome)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Бесполезные программы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into 'qgrs-cpp'...\n",
      "remote: Enumerating objects: 75, done.\u001b[K\n",
      "remote: Total 75 (delta 0), reused 0 (delta 0), pack-reused 75\u001b[K\n",
      "Unpacking objects: 100% (75/75), 281.53 KiB | 360.00 KiB/s, done.\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "! git clone https://github.com/freezer333/qgrs-cpp\n",
    "! g++ -o qgrs qgrs-cpp/src/default.cpp qgrs-cpp/src/qgrs.cpp --std=c++11\n",
    "! rm -rf qgrs-cpp\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>T1</th>\n",
       "      <th>T2</th>\n",
       "      <th>T3</th>\n",
       "      <th>T4</th>\n",
       "      <th>TS</th>\n",
       "      <th>GS</th>\n",
       "      <th>SEQ</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>28</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>GGACGGCGAGGTGCCCACTGGCGACGCAGG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    T1  T2  T3  T4  TS  GS                               SEQ\n",
       "ID                                                          \n",
       "1    0   9  19  28   2  20    GGACGGCGAGGTGCCCACTGGCGACGCAGG"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Не хватает оперативки\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from subprocess import DEVNULL, PIPE\n",
    "\n",
    "Q_EXECUTABLE = Path(\"../hse22_project/qgrs\")\n",
    "assert Q_EXECUTABLE.is_file()\n",
    "\n",
    "\n",
    "def qgrs(query: str):\n",
    "    assert set(query).issubset({\"A\", \"C\", \"G\", \"T\", \"N\"})\n",
    "    fd, temp = tempfile.mkstemp()\n",
    "    os.close(fd)\n",
    "    with open(temp, 'w') as stream:\n",
    "        stream.write(query)\n",
    "\n",
    "    subprocess.run(\n",
    "        [Q_EXECUTABLE, \"-csv\", \"-i\", temp, \"-o\", temp + \".qgrs\"],\n",
    "        check=True, stdout=PIPE, stderr=DEVNULL,\n",
    "        input=query, encoding='ascii'\n",
    "    )\n",
    "    \n",
    "    with open(temp + \".qgrs\", 'r') as stream:\n",
    "        df = pd.read_csv(stream,\n",
    "                         #names=['ID', 'T1', 'T2', 'T3', 'T4', 'TS', 'GS', 'SEQ'],\n",
    "                         skiprows=1, sep=',').iloc[:,:8]\n",
    "    os.remove(temp)\n",
    "    os.remove(temp + \".qgrs\")\n",
    "    df.set_index(\"ID\", inplace=True)\n",
    "    #df = df.loc[df[\"ZH-Score\"] > 500]\n",
    "    return df\n",
    "    # return df[['Start', 'End', 'ZH-Score', 'Sequence', 'Conformation']]\n",
    "\n",
    "qgrs(\"GGACGGCGAGGTGCCCACTGGCGACGCAGG\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "ename": "CalledProcessError",
     "evalue": "Command '[PosixPath('../hse22_project/qgrs'), '-csv', '-i', '/tmp/tmpzyrkgxy6', '-o', '/tmp/tmpzyrkgxy6.qgrs']' died with <Signals.SIGKILL: 9>.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[1;32m/home/narek/hse22_project/notebook.ipynb Cell 57'\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000079vscode-remote?line=0'>1</a>\u001b[0m qgrs(na_seqs[genome][:\u001b[39m500000\u001b[39;49m])\n",
      "\u001b[1;32m/home/narek/hse22_project/notebook.ipynb Cell 56'\u001b[0m in \u001b[0;36mqgrs\u001b[0;34m(query)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=16'>17</a>\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(temp, \u001b[39m'\u001b[39m\u001b[39mw\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mas\u001b[39;00m stream:\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=17'>18</a>\u001b[0m     stream\u001b[39m.\u001b[39mwrite(query)\n\u001b[0;32m---> <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=19'>20</a>\u001b[0m subprocess\u001b[39m.\u001b[39;49mrun(\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=20'>21</a>\u001b[0m     [Q_EXECUTABLE, \u001b[39m\"\u001b[39;49m\u001b[39m-csv\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39m-i\u001b[39;49m\u001b[39m\"\u001b[39;49m, temp, \u001b[39m\"\u001b[39;49m\u001b[39m-o\u001b[39;49m\u001b[39m\"\u001b[39;49m, temp \u001b[39m+\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39m.qgrs\u001b[39;49m\u001b[39m\"\u001b[39;49m],\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=21'>22</a>\u001b[0m     check\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m, stdout\u001b[39m=\u001b[39;49mPIPE, stderr\u001b[39m=\u001b[39;49mDEVNULL,\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=22'>23</a>\u001b[0m     \u001b[39minput\u001b[39;49m\u001b[39m=\u001b[39;49mquery, encoding\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mascii\u001b[39;49m\u001b[39m'\u001b[39;49m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=23'>24</a>\u001b[0m )\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=25'>26</a>\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(temp \u001b[39m+\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m.qgrs\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mr\u001b[39m\u001b[39m'\u001b[39m) \u001b[39mas\u001b[39;00m stream:\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=26'>27</a>\u001b[0m     df \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(stream,\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=27'>28</a>\u001b[0m                      \u001b[39m#names=['ID', 'T1', 'T2', 'T3', 'T4', 'TS', 'GS', 'SEQ'],\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell://wsl%2Bubuntu/home/narek/hse22_project/notebook.ipynb#ch0000054vscode-remote?line=28'>29</a>\u001b[0m                      skiprows\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m, sep\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39m,\u001b[39m\u001b[39m'\u001b[39m)\u001b[39m.\u001b[39miloc[:,:\u001b[39m8\u001b[39m]\n",
      "File \u001b[0;32m/usr/lib/python3.8/subprocess.py:516\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    <a href='file:///usr/lib/python3.8/subprocess.py?line=513'>514</a>\u001b[0m     retcode \u001b[39m=\u001b[39m process\u001b[39m.\u001b[39mpoll()\n\u001b[1;32m    <a href='file:///usr/lib/python3.8/subprocess.py?line=514'>515</a>\u001b[0m     \u001b[39mif\u001b[39;00m check \u001b[39mand\u001b[39;00m retcode:\n\u001b[0;32m--> <a href='file:///usr/lib/python3.8/subprocess.py?line=515'>516</a>\u001b[0m         \u001b[39mraise\u001b[39;00m CalledProcessError(retcode, process\u001b[39m.\u001b[39margs,\n\u001b[1;32m    <a href='file:///usr/lib/python3.8/subprocess.py?line=516'>517</a>\u001b[0m                                  output\u001b[39m=\u001b[39mstdout, stderr\u001b[39m=\u001b[39mstderr)\n\u001b[1;32m    <a href='file:///usr/lib/python3.8/subprocess.py?line=517'>518</a>\u001b[0m \u001b[39mreturn\u001b[39;00m CompletedProcess(process\u001b[39m.\u001b[39margs, retcode, stdout, stderr)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command '[PosixPath('../hse22_project/qgrs'), '-csv', '-i', '/tmp/tmpzyrkgxy6', '-o', '/tmp/tmpzyrkgxy6.qgrs']' died with <Signals.SIGKILL: 9>."
     ]
    }
   ],
   "source": [
    "qgrs(na_seqs[genome][:500000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! wget https://raw.githubusercontent.com/AnimaTardeb/G4Hunter/master/G4Hunter.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! wget https://bootstrap.pypa.io/pip/2.7/get-pip.py\n",
    "! python2 get-pip.py\n",
    "! rm get-pip.py\n",
    "! sudo apt-get install python2.7-tk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! python2 -m pip install numpy matplotlib biopython==1.76\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Не хватает оперативки\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from subprocess import DEVNULL, PIPE\n",
    "\n",
    "G4_EXECUTABLE = Path(\"G4Hunter.py\")\n",
    "assert G4_EXECUTABLE.is_file()\n",
    "\n",
    "\n",
    "def G4(query: str, windowsize: int = 6, threshold: int = 10):\n",
    "    assert set(query).issubset({\"A\", \"C\", \"G\", \"T\", \"N\"})\n",
    "    fd, temp = tempfile.mkstemp()\n",
    "    os.close(fd)\n",
    "    with open(temp, 'w') as stream:\n",
    "        stream.write(query)\n",
    "\n",
    "    subprocess.run(\n",
    "        [\"python2\", G4_EXECUTABLE, \"-i\", temp, \"-o\", \"test\" + \".G4\",\n",
    "         \"-w\", windowsize, \"-s\", threshold],\n",
    "        check=True, stdout=PIPE, stderr=DEVNULL,\n",
    "        input=query, encoding='ascii'\n",
    "    )\n",
    "\n",
    "\n",
    "    '''\n",
    "    with open(temp + \".qgrs\", 'r') as stream:\n",
    "        df = pd.read_csv(stream,\n",
    "                         names=['ID, 'T1', 'T2', 'T3', 'T4', 'TS', 'GS', 'SEQ'],\n",
    "                         skiprows=1, sep='\\s+')\n",
    "    os.remove(temp)\n",
    "    os.remove(temp + \".Z-SCORE\")\n",
    "    df = df.loc[df[\"ZH-Score\"] > 500]\n",
    "    return df[['Start', 'End', 'ZH-Score', 'Sequence', 'Conformation']]\n",
    "    '''\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ставим PQSfinder и запускаем"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! Rscript configure.r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>score</th>\n",
       "      <th>nt</th>\n",
       "      <th>nb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr</td>\n",
       "      <td>81343</td>\n",
       "      <td>81362</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>66</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr</td>\n",
       "      <td>123220</td>\n",
       "      <td>123236</td>\n",
       "      <td>17</td>\n",
       "      <td>+</td>\n",
       "      <td>70</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr</td>\n",
       "      <td>142561</td>\n",
       "      <td>142580</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>66</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr</td>\n",
       "      <td>190647</td>\n",
       "      <td>190667</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>64</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr</td>\n",
       "      <td>190963</td>\n",
       "      <td>190987</td>\n",
       "      <td>25</td>\n",
       "      <td>+</td>\n",
       "      <td>59</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr</td>\n",
       "      <td>325970</td>\n",
       "      <td>325993</td>\n",
       "      <td>24</td>\n",
       "      <td>-</td>\n",
       "      <td>60</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr</td>\n",
       "      <td>338220</td>\n",
       "      <td>338248</td>\n",
       "      <td>29</td>\n",
       "      <td>-</td>\n",
       "      <td>54</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr</td>\n",
       "      <td>475142</td>\n",
       "      <td>475164</td>\n",
       "      <td>23</td>\n",
       "      <td>-</td>\n",
       "      <td>61</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr</td>\n",
       "      <td>494361</td>\n",
       "      <td>494378</td>\n",
       "      <td>18</td>\n",
       "      <td>+</td>\n",
       "      <td>69</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr</td>\n",
       "      <td>540475</td>\n",
       "      <td>540494</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>66</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr</td>\n",
       "      <td>633837</td>\n",
       "      <td>633859</td>\n",
       "      <td>23</td>\n",
       "      <td>+</td>\n",
       "      <td>61</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr</td>\n",
       "      <td>714240</td>\n",
       "      <td>714265</td>\n",
       "      <td>26</td>\n",
       "      <td>+</td>\n",
       "      <td>57</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr</td>\n",
       "      <td>751833</td>\n",
       "      <td>751854</td>\n",
       "      <td>22</td>\n",
       "      <td>+</td>\n",
       "      <td>63</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr</td>\n",
       "      <td>933659</td>\n",
       "      <td>933688</td>\n",
       "      <td>30</td>\n",
       "      <td>+</td>\n",
       "      <td>52</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr</td>\n",
       "      <td>983156</td>\n",
       "      <td>983179</td>\n",
       "      <td>24</td>\n",
       "      <td>+</td>\n",
       "      <td>60</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqnames   start     end  width strand  score  nt  nb\n",
       "0       chr   81343   81362     20      +     66   3   0\n",
       "1       chr  123220  123236     17      +     70   3   0\n",
       "2       chr  142561  142580     20      +     66   3   0\n",
       "3       chr  190647  190667     21      +     64   3   0\n",
       "4       chr  190963  190987     25      +     59   3   0\n",
       "5       chr  325970  325993     24      -     60   3   0\n",
       "6       chr  338220  338248     29      -     54   3   0\n",
       "7       chr  475142  475164     23      -     61   3   0\n",
       "8       chr  494361  494378     18      +     69   3   0\n",
       "9       chr  540475  540494     20      +     66   3   0\n",
       "10      chr  633837  633859     23      +     61   3   0\n",
       "11      chr  714240  714265     26      +     57   3   0\n",
       "12      chr  751833  751854     22      +     63   3   0\n",
       "13      chr  933659  933688     30      +     52   3   0\n",
       "14      chr  983156  983179     24      +     60   3   0"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "from subprocess import DEVNULL, PIPE\n",
    "\n",
    "Q_EXECUTABLE = Path(\"pqsfinder.r\")\n",
    "assert Q_EXECUTABLE.is_file()\n",
    "\n",
    "\n",
    "def pqs(query: str):\n",
    "    assert set(query).issubset({\"A\", \"C\", \"G\", \"T\", \"N\"})\n",
    "    fd, temp = tempfile.mkstemp()\n",
    "    os.close(fd)\n",
    "    with open(temp, 'w') as stream:\n",
    "        stream.write(\">chr\\n\"+query)\n",
    "\n",
    "    subprocess.run(\n",
    "        [\"Rscript\", Q_EXECUTABLE, temp, temp + \".qgrs\"],\n",
    "        check=True, stdout=PIPE, stderr=DEVNULL,\n",
    "        input=query, encoding='ascii'\n",
    "    )\n",
    "    \n",
    "    with open(temp + \".qgrs\", 'r') as stream:\n",
    "        df = pd.read_csv(stream,\n",
    "                         #names=['ID', 'T1', 'T2', 'T3', 'T4', 'TS', 'GS', 'SEQ'],\n",
    "                         skiprows=0, sep=' ').iloc[:,:8]\n",
    "    os.remove(temp)\n",
    "    os.remove(temp + \".qgrs\")\n",
    "    #df.set_index(\"ID\", inplace=True)\n",
    "    #df = df.loc[df[\"ZH-Score\"] > 500]\n",
    "    return df\n",
    "    # return df[['Start', 'End', 'ZH-Score', 'Sequence', 'Conformation']]\n",
    "\n",
    "pqs(na_seqs[genome][:1000000])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.10 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
